<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Hierarchical Linear Regression (Radon Contamination dataset) – Bambi</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../notebooks/Strack_RRR_re_analysis.html" rel="next">
<link href="../notebooks/sleepstudy.html" rel="prev">
<link href="../logos/favicon.png" rel="icon" type="image/png">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@*/dist/embed-amd.js" crossorigin="anonymous"></script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../logos/RGB/Bambi_logo.png" alt="bambi-home" class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../notebooks/getting_started.html"> 
<span class="menu-text">Getting Started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../api/index.html"> 
<span class="menu-text">API Reference</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="../notebooks/index.html" aria-current="page"> 
<span class="menu-text">Examples</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../faq.html"> 
<span class="menu-text">FAQ</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../changelog.html"> 
<span class="menu-text">Changelog</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../contact.html"> 
<span class="menu-text">Contact</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/bambinos/bambi"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../notebooks/t-test.html">Linear regression models</a></li><li class="breadcrumb-item"><a href="../notebooks/radon_example.html">Hierarchical Linear Regression (Radon Contamination dataset)</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Examples gallery</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Linear regression models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/t-test.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Comparison of two means (T-test)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/ESCS_multiple_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multiple linear regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/splines_cherry_blossoms.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Regression splines (Cherry blossom example)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/multi-level_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical Linear Regression (Pigs dataset)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/sleepstudy.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical Linear Regression (Sleepstudy example)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/radon_example.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Hierarchical Linear Regression (Radon Contamination dataset)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/Strack_RRR_re_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bayesian Workflow (Strack RRR Analysis Replication)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/shooter_crossed_random_ANOVA.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bayesian Workflow (Police Officer’s Dilemma)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/t_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Robust Linear Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/predict_new_groups.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Predict New Groups</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/polynomial_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Polynomial Regression</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false">
 <span class="menu-text">Generalized linear models</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/logistic_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Logistic Regression (Vote intention with ANES data)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/model_comparison.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Logistic Regression and Model Comparison with Bambi and ArviZ</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/hierarchical_binomial_bambi.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Hierarchical Logistic regression with Binomial family</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/alternative_links_binary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Regression for Binary responses: Alternative link functions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/wald_gamma_glm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Wald and Gamma Regression (Australian insurance claims 2004-2005)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/negative_binomial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Negative Binomial Regression (Students absence example)</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/count_roaches.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Count Regression with Variable Exposure</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/beta_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Beta Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/categorical_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Categorical Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/circular_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Circular Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/quantile_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quantile Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/mister_p.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Multilevel Regression and Post-stratification</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/ordinal_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ordinal Regression</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/zero_inflated_regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Zero inflated models</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false">
 <span class="menu-text">More advanced models</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/distributional_models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Distributional models</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/hsgp_1d.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gaussian Processes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/hsgp_2d.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gaussian Processes in 2D</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/survival_model.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Survival Models</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/orthogonal_polynomial_reg.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Orthogonal Polynomial regression</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false">
 <span class="menu-text">Tools to interpret model outputs</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/plot_predictions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plot Conditional Adjusted Predictions</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/plot_comparisons.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plot Comparisons</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/plot_slopes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Plot Slopes</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/interpret_advanced_usage.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Interpret Advanced Usage</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false">
 <span class="menu-text">Alternative sampling backends</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../notebooks/alternative_samplers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Alternative sampling backends</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#prepare-notebook" id="toc-prepare-notebook" class="nav-link active" data-scroll-target="#prepare-notebook">Prepare Notebook</a></li>
  <li><a href="#read-data" id="toc-read-data" class="nav-link" data-scroll-target="#read-data">Read Data</a></li>
  <li><a href="#preprocess-data" id="toc-preprocess-data" class="nav-link" data-scroll-target="#preprocess-data">Preprocess Data</a></li>
  <li><a href="#eda" id="toc-eda" class="nav-link" data-scroll-target="#eda">EDA</a></li>
  <li><a href="#models-conventional-approaches" id="toc-models-conventional-approaches" class="nav-link" data-scroll-target="#models-conventional-approaches">Models: Conventional Approaches</a>
  <ul class="collapse">
  <li><a href="#complete-pooling" id="toc-complete-pooling" class="nav-link" data-scroll-target="#complete-pooling">Complete Pooling</a></li>
  <li><a href="#no-pooling" id="toc-no-pooling" class="nav-link" data-scroll-target="#no-pooling">No Pooling</a></li>
  </ul></li>
  <li><a href="#multilevel-and-hierarchical-models" id="toc-multilevel-and-hierarchical-models" class="nav-link" data-scroll-target="#multilevel-and-hierarchical-models">Multilevel and hierarchical models</a>
  <ul class="collapse">
  <li><a href="#partial-pooling-model" id="toc-partial-pooling-model" class="nav-link" data-scroll-target="#partial-pooling-model">Partial pooling model</a></li>
  <li><a href="#county-specific-intercepts-and-common-predictors" id="toc-county-specific-intercepts-and-common-predictors" class="nav-link" data-scroll-target="#county-specific-intercepts-and-common-predictors">County-specific intercepts and common predictors</a></li>
  <li><a href="#varying-intercept-and-slope-model" id="toc-varying-intercept-and-slope-model" class="nav-link" data-scroll-target="#varying-intercept-and-slope-model">Varying intercept and slope model</a></li>
  <li><a href="#adding-group-level-predictors" id="toc-adding-group-level-predictors" class="nav-link" data-scroll-target="#adding-group-level-predictors">Adding group-level predictors</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../notebooks/t-test.html">Linear regression models</a></li><li class="breadcrumb-item"><a href="../notebooks/radon_example.html">Hierarchical Linear Regression (Radon Contamination dataset)</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">Hierarchical Linear Regression (Radon Contamination dataset)</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>In this notebook we want to revisit the classical hierarchical linear regression model based on the dataset of the <em>Radon Contamination</em> by Gelman and Hill. In particular, we want to show how easy is to port the <a href="https://docs.pymc.io/en/v3/">PyMC</a> models, presented in the very complete article <a href="https://docs.pymc.io/projects/examples/en/latest/case_studies/multilevel_modeling.html">A Primer on Bayesian Methods for Multilevel Modeling</a>, to <a href="https://bambinos.github.io/bambi/notebooks/">Bambi</a> using the more concise formula specification for the models.</p>
<p>This example has been ported from PyMC by Juan Orduz (<a href="https://github.com/juanitorduz">@juanitorduz</a>) and Bambi developers.</p>
<section id="prepare-notebook" class="level2">
<h2 class="anchored" data-anchor-id="prepare-notebook">Prepare Notebook</h2>
<div id="cell-3" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> arviz <span class="im">as</span> az</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> bambi <span class="im">as</span> bmb</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pymc <span class="im">as</span> pm</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-4" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>az.style.use(<span class="st">"arviz-darkgrid"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>np.random.default_rng(<span class="dv">8924</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="2">
<pre><code>Generator(PCG64) at 0x76AD7DB7DE00</code></pre>
</div>
</div>
</section>
<section id="read-data" class="level2">
<h2 class="anchored" data-anchor-id="read-data">Read Data</h2>
<p>Let us load the data into a pandas data frame.</p>
<div id="cell-6" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get radon data</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/pymc-devs/pymc-examples/main/examples/data/srrs2.dat"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>radon_df <span class="op">=</span> pd.read_csv(path)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Get city data</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>city_df <span class="op">=</span> pd.read_csv(pm.get_data(<span class="st">"cty.dat"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-7" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>display(radon_df.head())</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(radon_df.shape[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">idnum</th>
<th data-quarto-table-cell-role="th">state</th>
<th data-quarto-table-cell-role="th">state2</th>
<th data-quarto-table-cell-role="th">stfips</th>
<th data-quarto-table-cell-role="th">zip</th>
<th data-quarto-table-cell-role="th">region</th>
<th data-quarto-table-cell-role="th">typebldg</th>
<th data-quarto-table-cell-role="th">floor</th>
<th data-quarto-table-cell-role="th">room</th>
<th data-quarto-table-cell-role="th">basement</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">stoptm</th>
<th data-quarto-table-cell-role="th">startdt</th>
<th data-quarto-table-cell-role="th">stopdt</th>
<th data-quarto-table-cell-role="th">activity</th>
<th data-quarto-table-cell-role="th">pcterr</th>
<th data-quarto-table-cell-role="th">adjwt</th>
<th data-quarto-table-cell-role="th">dupflag</th>
<th data-quarto-table-cell-role="th">zipflag</th>
<th data-quarto-table-cell-role="th">cntyfips</th>
<th data-quarto-table-cell-role="th">county</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>1</td>
<td>AZ</td>
<td>AZ</td>
<td>4</td>
<td>85920</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>2</td>
<td>N</td>
<td>...</td>
<td>1100</td>
<td>112987</td>
<td>120287</td>
<td>0.3</td>
<td>0.0</td>
<td>136.060971</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>APACHE</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2</td>
<td>AZ</td>
<td>AZ</td>
<td>4</td>
<td>85920</td>
<td>1</td>
<td>0</td>
<td>9</td>
<td>0</td>
<td></td>
<td>...</td>
<td>700</td>
<td>70788</td>
<td>71188</td>
<td>0.6</td>
<td>33.3</td>
<td>128.784975</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>APACHE</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>3</td>
<td>AZ</td>
<td>AZ</td>
<td>4</td>
<td>85924</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>3</td>
<td>N</td>
<td>...</td>
<td>1145</td>
<td>70788</td>
<td>70788</td>
<td>0.5</td>
<td>0.0</td>
<td>150.245112</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>APACHE</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4</td>
<td>AZ</td>
<td>AZ</td>
<td>4</td>
<td>85925</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>3</td>
<td>N</td>
<td>...</td>
<td>1900</td>
<td>52088</td>
<td>52288</td>
<td>0.6</td>
<td>97.2</td>
<td>136.060971</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>APACHE</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5</td>
<td>AZ</td>
<td>AZ</td>
<td>4</td>
<td>85932</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>N</td>
<td>...</td>
<td>900</td>
<td>70788</td>
<td>70788</td>
<td>0.3</td>
<td>0.0</td>
<td>136.060971</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>APACHE</td>
</tr>
</tbody>
</table>

<p>5 rows × 25 columns</p>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>12777</code></pre>
</div>
</div>
<div id="cell-8" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>display(city_df.head())</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(city_df.shape[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">stfips</th>
<th data-quarto-table-cell-role="th">ctfips</th>
<th data-quarto-table-cell-role="th">st</th>
<th data-quarto-table-cell-role="th">cty</th>
<th data-quarto-table-cell-role="th">lon</th>
<th data-quarto-table-cell-role="th">lat</th>
<th data-quarto-table-cell-role="th">Uppm</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>1</td>
<td>1</td>
<td>AL</td>
<td>AUTAUGA</td>
<td>-86.643</td>
<td>32.534</td>
<td>1.78331</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1</td>
<td>3</td>
<td>AL</td>
<td>BALDWIN</td>
<td>-87.750</td>
<td>30.661</td>
<td>1.38323</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>1</td>
<td>5</td>
<td>AL</td>
<td>BARBOUR</td>
<td>-85.393</td>
<td>31.870</td>
<td>2.10105</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>1</td>
<td>7</td>
<td>AL</td>
<td>BIBB</td>
<td>-87.126</td>
<td>32.998</td>
<td>1.67313</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1</td>
<td>9</td>
<td>AL</td>
<td>BLOUNT</td>
<td>-86.568</td>
<td>33.981</td>
<td>1.88501</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>3194</code></pre>
</div>
</div>
</section>
<section id="preprocess-data" class="level2">
<h2 class="anchored" data-anchor-id="preprocess-data">Preprocess Data</h2>
<p>We are going to preprocess the data as done in the article <a href="https://docs.pymc.io/projects/examples/en/latest/case_studies/multilevel_modeling.html">A Primer on Bayesian Methods for Multilevel Modeling</a>.</p>
<div id="cell-10" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Strip spaces from column names</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>radon_df.columns <span class="op">=</span> radon_df.columns.<span class="bu">map</span>(<span class="bu">str</span>.strip)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter to keep observations for "MN" state only</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> radon_df[radon_df.state <span class="op">==</span> <span class="st">"MN"</span>].copy()</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>city_mn_df <span class="op">=</span> city_df[city_df.st <span class="op">==</span> <span class="st">"MN"</span>].copy()</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute fips</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"fips"</span>] <span class="op">=</span> <span class="dv">1_000</span> <span class="op">*</span> df.stfips <span class="op">+</span> df.cntyfips</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>city_mn_df[<span class="st">"fips"</span>] <span class="op">=</span> <span class="dv">1_000</span> <span class="op">*</span> city_mn_df.stfips <span class="op">+</span> city_mn_df.ctfips</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge data</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.merge(city_mn_df[[<span class="st">"fips"</span>, <span class="st">"Uppm"</span>]], on<span class="op">=</span><span class="st">"fips"</span>)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.drop_duplicates(subset<span class="op">=</span><span class="st">"idnum"</span>)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Clean county names</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>df.county <span class="op">=</span> df.county.<span class="bu">map</span>(<span class="bu">str</span>.strip)</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute log(radon + 0.1)</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"log_radon"</span>] <span class="op">=</span> np.log(df[<span class="st">"activity"</span>] <span class="op">+</span> <span class="fl">0.1</span>)</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute log of Uranium</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"log_u"</span>] <span class="op">=</span> np.log(df[<span class="st">"Uppm"</span>])</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's map floor. 0 -&gt; Basement and 1 -&gt; Floor</span></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"floor"</span>] <span class="op">=</span> df[<span class="st">"floor"</span>].<span class="bu">map</span>({<span class="dv">0</span>: <span class="st">"Basement"</span>, <span class="dv">1</span>: <span class="st">"Floor"</span>})</span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-28"><a href="#cb9-28" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort values by floor</span></span>
<span id="cb9-29"><a href="#cb9-29" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.sort_values(by<span class="op">=</span><span class="st">"floor"</span>)</span>
<span id="cb9-30"><a href="#cb9-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-31"><a href="#cb9-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Reset index</span></span>
<span id="cb9-32"><a href="#cb9-32" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this exercise, we model the logarithm of the Radon measurements. This is because the distribution of the Radon level is approximately log-normal. We also add a small number, 0.1, to prevent us from trying to compute the logarithm of 0.</p>
</section>
<section id="eda" class="level2">
<h2 class="anchored" data-anchor-id="eda">EDA</h2>
<p>In order to get a glimpse of the data, we are going to do some exploratory data analysis. First, let’s have a look at the global distribution of the untransformed radon levels.</p>
<div id="cell-13" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>_, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>sns.histplot(x<span class="op">=</span><span class="st">"activity"</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, stat<span class="op">=</span><span class="st">"density"</span>, element<span class="op">=</span><span class="st">"step"</span>, common_norm<span class="op">=</span><span class="va">False</span>, data<span class="op">=</span>df, ax<span class="op">=</span>ax)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>sns.kdeplot(x<span class="op">=</span><span class="st">"activity"</span>, data<span class="op">=</span>df, ax<span class="op">=</span>ax, cut<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Density of Radon"</span>, xlabel<span class="op">=</span><span class="st">"Radon"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-8-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Next, let us see the global <code>log(radon + 0.1)</code> distribution.</p>
<div id="cell-15" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>_, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sns.histplot(x<span class="op">=</span><span class="st">"log_radon"</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, stat<span class="op">=</span><span class="st">"density"</span>, element<span class="op">=</span><span class="st">"step"</span>, common_norm<span class="op">=</span><span class="va">False</span>, data<span class="op">=</span>df, ax<span class="op">=</span>ax)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>sns.kdeplot(x<span class="op">=</span><span class="st">"log_radon"</span>, data<span class="op">=</span>df, ax<span class="op">=</span>ax)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Density of log(Radon + 0.1)"</span>, xlabel<span class="op">=</span><span class="st">"$\log(Radon + 0.1)$"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-9-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>There are many a priori reasons to think houses with basement has higher radon levels. From geochemistry to composition of building materials to poor ventilation. We can split the distribution of <code>log(radon + 0.1)</code> per <code>floor</code> to see if we are able to see that difference in our data.</p>
<div id="cell-17" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>_, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>sns.histplot(</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"log_radon"</span>, hue<span class="op">=</span><span class="st">"floor"</span>, alpha<span class="op">=</span><span class="fl">0.2</span>, stat<span class="op">=</span><span class="st">"density"</span>, element<span class="op">=</span><span class="st">"step"</span>, </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    common_norm<span class="op">=</span><span class="va">False</span>, data<span class="op">=</span>df, ax<span class="op">=</span>ax</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>sns.kdeplot(x<span class="op">=</span><span class="st">"log_radon"</span>, hue<span class="op">=</span><span class="st">"floor"</span>, common_norm<span class="op">=</span><span class="va">False</span>, data<span class="op">=</span>df, ax<span class="op">=</span>ax)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Density of log(Radon + 0.1)"</span>, xlabel<span class="op">=</span><span class="st">"$\log + 0.1$"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-10-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This exploration tell us that, as expected, the average radon level is higher in <code>Basement</code> than <code>Floor</code>.</p>
<p>Next, we are going to count the number of counties.</p>
<div id="cell-20" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>n_counties <span class="op">=</span> df[<span class="st">"county"</span>].unique().size</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of counties: </span><span class="sc">{</span>n_counties<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of counties: 85</code></pre>
</div>
</div>
<p>Let us dig deeper into the distribution of radon and number of observations per county and floor level.</p>
<div id="cell-22" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>log_radon_county_agg  <span class="op">=</span> (</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    df </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    .groupby([<span class="st">"county"</span>, <span class="st">"floor"</span>], as_index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    .agg(</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>        log_radon_mean<span class="op">=</span>(<span class="st">"log_radon"</span>, <span class="st">"mean"</span>),</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>        n_obs<span class="op">=</span>(<span class="st">"log_radon"</span>, <span class="st">"count"</span>)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>fig, ax<span class="op">=</span> plt.subplots(nrows<span class="op">=</span><span class="dv">1</span>, ncols<span class="op">=</span><span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">6</span>), layout<span class="op">=</span><span class="st">"constrained"</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>sns.boxplot(x<span class="op">=</span><span class="st">"floor"</span>, y<span class="op">=</span><span class="st">"log_radon_mean"</span>, data<span class="op">=</span>log_radon_county_agg, ax<span class="op">=</span>ax[<span class="dv">0</span>])</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>ax[<span class="dv">0</span>].<span class="bu">set</span>(title<span class="op">=</span><span class="st">"log(Radon + 0.1) Mean per County"</span>, ylabel<span class="op">=</span><span class="st">"$\log + 0.1$"</span>)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>sns.boxplot(x<span class="op">=</span><span class="st">"floor"</span>, y<span class="op">=</span><span class="st">"n_obs"</span>, data<span class="op">=</span>log_radon_county_agg, ax<span class="op">=</span>ax[<span class="dv">1</span>])</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>ax[<span class="dv">1</span>].<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Number of Observations"</span>, xlabel<span class="op">=</span><span class="st">"floor"</span>, ylabel<span class="op">=</span><span class="st">"Number of observations"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>On the left hand side we can see that the <code>"Basement"</code> distribution per county is shifted to higher values with respect to the <code>"Floor"</code> distribution. We had seen this above when considering all counties together.</li>
<li>On the right hand side we see that the number of observations per county is not the same for the <code>floor</code> levels. In particular, we see that there are some counties with a lot of basement observations. This can create some bias when computing simple statistics to compare across counties. Moreover, not all <code>county</code> and <code>floor</code> combinations are present in the dataset. For example:</li>
</ul>
<div id="cell-24" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> df.query(<span class="st">"county == 'YELLOW MEDICINE' and floor == 'Floor'"</span>).empty</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="models-conventional-approaches" class="level2">
<h2 class="anchored" data-anchor-id="models-conventional-approaches">Models: Conventional Approaches</h2>
<section id="complete-pooling" class="level3">
<h3 class="anchored" data-anchor-id="complete-pooling">Complete Pooling</h3>
<section id="model" class="level4">
<h4 class="anchored" data-anchor-id="model">Model</h4>
<p>For this first model we only consider the predictor <code>floor</code>, which represents the floor level. The following equation describes the linear model that we are going to build with Bambi</p>
<p><span class="math display">\[
y = \beta_{j} + \varepsilon
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
y &amp;= \text{Response for the (log) radon measurement }\\
\beta_{j} &amp;= \text{Coefficient for the floor level } j \\
\varepsilon &amp; = \text{Residual random error}
\end{aligned}
\]</span></p>
<p>Each <span class="math inline">\(j\)</span> indexes a different floor level. In this case, <span class="math inline">\(j=1\)</span> means <code>"basement"</code> and <span class="math inline">\(j=2\)</span> means <code>"floor"</code>.</p>
</section>
<section id="priors" class="level4">
<h4 class="anchored" data-anchor-id="priors">Priors</h4>
<section id="common-effects" class="level5">
<h5 class="anchored" data-anchor-id="common-effects">Common effects</h5>
<p>The only common effect in this model is the floor effect represented by the <span class="math inline">\(\beta_{j}\)</span> coefficients. We have</p>
<p><span class="math display">\[
\beta_{j} \sim \text{Normal}(0, \sigma_{\beta_j})
\]</span></p>
<p>for <span class="math inline">\(j: 1, 2\)</span>, where <span class="math inline">\(\sigma_{\beta_j}\)</span> is a positive constant that we set to 10 for all <span class="math inline">\(j\)</span>.</p>
</section>
<section id="residual-error" class="level5">
<h5 class="anchored" data-anchor-id="residual-error">Residual error</h5>
<p><span class="math display">\[
\begin{aligned}
\varepsilon &amp; \sim \text{Normal}(0, \sigma) \\
\sigma &amp; \sim \text{Exponential}(\lambda)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\lambda\)</span> is a positive constant that we set to 1.</p>
<p>Let us now write the Bambi model.</p>
<p>The <code>0</code> on the right side of <code>~</code> in the model formula removes the global intercept that is added by default. This allows Bambi to use one coefficient for each floor level.</p>
<div id="cell-30" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># A dictionary with the priors we pass to the model initialization</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>pooled_priors <span class="op">=</span> {</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"floor"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sigma"</span>: bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>pooled_model <span class="op">=</span> bmb.Model(<span class="st">"log_radon ~ 0 + floor"</span>, df, priors<span class="op">=</span>pooled_priors)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>pooled_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>       Formula: log_radon ~ 0 + floor
        Family: gaussian
          Link: mu = identity
  Observations: 919
        Priors: 
    target = mu
        Common-level effects
            floor ~ Normal(mu: 0.0, sigma: 10.0)
        
        Auxiliary parameters
            sigma ~ Exponential(lam: 1.0)</code></pre>
</div>
</div>
<p>The <code>Family name: Gaussian</code> indicates the selected family, which defaults to Gaussian. And <code>Link: identity</code> indicates the default value for the link argument in <code>bmb.Model()</code>. Taken together this simply means that we are fitting a normal linear regression model.</p>
<p>Let’s see the graph representation of the model before fitting. To do so, we first need to call the <code>.build()</code> method. Internally, this builds the underlying PyMC model.</p>
<div id="cell-33" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pooled_model.build()</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>pooled_model.graph()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-15-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Let’s now fit the model.</p>
<div id="cell-35" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pooled_results <span class="op">=</span> pooled_model.fit()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [sigma, floor]</code></pre>
</div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"392ea9efbea3411680fe950031c84493","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sampling 2 chains for 1_000 tune and 1_000 draw iterations (2_000 + 2_000 draws total) took 3 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics</code></pre>
</div>
</div>
<p>Now we can examine the posterior distribution, i.e.&nbsp;the joint distribution of model parameters conditional on the data:</p>
<div id="cell-37" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>az.plot_trace(data<span class="op">=</span>pooled_results, compact<span class="op">=</span><span class="va">True</span>, chain_prop<span class="op">=</span>{<span class="st">"ls"</span>: <span class="st">"-"</span>})</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Pooled Model Trace"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-17-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can also see some posterior summary statistics.</p>
<div id="cell-39" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>pooled_summary <span class="op">=</span> az.summary(data<span class="op">=</span>pooled_results)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>pooled_summary</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">mean</th>
<th data-quarto-table-cell-role="th">sd</th>
<th data-quarto-table-cell-role="th">hdi_3%</th>
<th data-quarto-table-cell-role="th">hdi_97%</th>
<th data-quarto-table-cell-role="th">mcse_mean</th>
<th data-quarto-table-cell-role="th">mcse_sd</th>
<th data-quarto-table-cell-role="th">ess_bulk</th>
<th data-quarto-table-cell-role="th">ess_tail</th>
<th data-quarto-table-cell-role="th">r_hat</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">floor[Basement]</td>
<td>1.362</td>
<td>0.029</td>
<td>1.309</td>
<td>1.416</td>
<td>0.001</td>
<td>0.000</td>
<td>3280.0</td>
<td>1577.0</td>
<td>1.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">floor[Floor]</td>
<td>0.776</td>
<td>0.066</td>
<td>0.657</td>
<td>0.898</td>
<td>0.001</td>
<td>0.001</td>
<td>3035.0</td>
<td>1397.0</td>
<td>1.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">sigma</td>
<td>0.791</td>
<td>0.019</td>
<td>0.752</td>
<td>0.824</td>
<td>0.000</td>
<td>0.000</td>
<td>3143.0</td>
<td>1483.0</td>
<td>1.0</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>From the posterior plot and the summary, we can see the mean radon level is considerably higher in the Basement than in the Floor level. This reflects what we originally saw in the initial data exploration. In addition, sice we have more measurements in the Basement, the uncertainty in its posterior is smaller than the uncertainty in the posterior for the Floor level.</p>
<p>We can compare the mean of the posterior distribution of the <code>floor</code> terms to the sample mean. This is going to be useful to understand the meaning of <strong>complete pooling</strong>.</p>
<div id="cell-42" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>_, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    pooled_summary[<span class="st">"mean"</span>]</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>    .iloc[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    .reset_index()</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>    .assign(floor <span class="op">=</span> <span class="kw">lambda</span> x: x[<span class="st">"index"</span>].<span class="bu">str</span>.<span class="bu">slice</span>(<span class="dv">6</span>, <span class="op">-</span><span class="dv">1</span>).<span class="bu">str</span>.strip())</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>    .merge(</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>        right<span class="op">=</span>df.groupby([<span class="st">"floor"</span>])[<span class="st">"log_radon"</span>].mean(),</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>        left_on<span class="op">=</span><span class="st">"floor"</span>,</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>        right_index<span class="op">=</span><span class="va">True</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>        <span class="st">"mean"</span>: <span class="st">"posterior mean"</span>,</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>        <span class="st">"log_radon"</span>: <span class="st">"sample mean"</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>    .melt(</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>        id_vars<span class="op">=</span><span class="st">"floor"</span>,</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>        value_vars<span class="op">=</span>[<span class="st">"posterior mean"</span>, <span class="st">"sample mean"</span>]</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>    .pipe((sns.barplot, <span class="st">"data"</span>),</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>        x<span class="op">=</span><span class="st">"floor"</span>,</span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>        y<span class="op">=</span><span class="st">"value"</span>,</span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>        hue<span class="op">=</span><span class="st">"variable"</span>,</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>        ax<span class="op">=</span>ax</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(title<span class="op">=</span><span class="st">"log(Radon + 0.1) Mean per Floor - Pooled Model"</span>, ylabel<span class="op">=</span><span class="st">"$\log + 0.1$"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>From the plot alone it is hard to detect the difference between the posterior mean and the sample mean. This happens because the estimation for any observation in either group is simply the group mean plus the smoothing due to the non-flat priors.</p>
<p>In other words, for every observation where <code>floor</code> is <code>"Basement"</code> the model predicts the mean radon for all the basement measurements, and for every observation where <code>floor</code> is <code>"Floor"</code>, the model predicts the mean radon for all the floor measurements.</p>
<p><strong>What does <em>complete pooling</em> exactly mean here?</strong></p>
<p>In this example, the pooling refers to how we treat the different counties when computing estimates (i.e.&nbsp;this does not refer to pooling across floor levels for example). <em>Complete pooling</em> means that all measurements for all counties are pooled into a single estimate (“treat all counties the same”), conditional on the floor level (because it is used as a covariate/predictor). For that reason, when computing the prediction for a given observation, we do not discriminate which county it belongs to. We pool all the counties into a single estimate, or in other words, we perform a complete pooling.</p>
<p>Let’s now compare the posterior predictive distribution for each group with the distribution of the observed data.</p>
<p>To do this we need to perform a couple of steps:</p>
<ul>
<li>Obtain samples from the posterior predictive distribution using the <code>.predict()</code> method.</li>
<li>Apply the inverse transform to have the posterior predictive samples in the original scale of the response.</li>
</ul>
<div id="cell-45" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note we create a new data set. </span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># One observation per group is enough to obtain posterior predictive samples for that group</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The more observations we create, the more posterior predictive samples from the same distribution</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we obtain.</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>new_data <span class="op">=</span> pd.DataFrame({<span class="st">"floor"</span>: [<span class="st">"Basement"</span>, <span class="st">"Floor"</span>]})</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>pooled_model.predict(pooled_results, kind<span class="op">=</span><span class="st">"pps"</span>, data<span class="op">=</span>new_data)</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Stack chains and draws and extract posterior predictive samples</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>pps <span class="op">=</span> az.extract_dataset(pooled_results, group<span class="op">=</span><span class="st">"posterior_predictive"</span>)[<span class="st">"log_radon"</span>].values</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Inverse transform the posterior predictive samples</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>pps <span class="op">=</span> np.exp(pps) <span class="op">-</span> <span class="fl">0.1</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(nrows<span class="op">=</span><span class="dv">2</span>, ncols<span class="op">=</span><span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">12</span>, <span class="dv">6</span>), layout<span class="op">=</span><span class="st">"constrained"</span>)</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> ax.flatten()</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>sns.histplot(x<span class="op">=</span>pps[<span class="dv">0</span>].flatten(), stat<span class="op">=</span><span class="st">"density"</span>, color<span class="op">=</span><span class="st">"C0"</span>, ax<span class="op">=</span>ax[<span class="dv">0</span>])</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>ax[<span class="dv">0</span>].<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Basement (Posterior Predictive Distribution)"</span>, xlabel<span class="op">=</span><span class="st">"radon"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>sns.histplot(x<span class="op">=</span><span class="st">"activity"</span>, data<span class="op">=</span>df.query(<span class="st">"floor == 'Basement'"</span>), stat<span class="op">=</span><span class="st">"density"</span>, ax<span class="op">=</span>ax[<span class="dv">2</span>])</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>ax[<span class="dv">2</span>].<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Basement (Sample Distribution)"</span>, xlim<span class="op">=</span>ax[<span class="dv">0</span>].get_xlim(), xlabel<span class="op">=</span><span class="st">"radon"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>sns.histplot(x<span class="op">=</span>pps[<span class="dv">1</span>].flatten(), stat<span class="op">=</span><span class="st">"density"</span>, color<span class="op">=</span><span class="st">"C1"</span>, ax<span class="op">=</span>ax[<span class="dv">1</span>])</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>ax[<span class="dv">1</span>].<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Floor (Posterior Predictive Distribution)"</span>, xlabel<span class="op">=</span><span class="st">"radon"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>sns.histplot(x<span class="op">=</span><span class="st">"activity"</span>, data<span class="op">=</span>df.query(<span class="st">"floor == 'Floor'"</span>), stat<span class="op">=</span><span class="st">"density"</span>, color<span class="op">=</span><span class="st">"C1"</span>, ax<span class="op">=</span>ax[<span class="dv">3</span>])</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a>ax[<span class="dv">3</span>].<span class="bu">set</span>(title<span class="op">=</span><span class="st">"Floor (Sample Distribution)"</span>, xlim<span class="op">=</span>ax[<span class="dv">1</span>].get_xlim(), xlabel<span class="op">=</span><span class="st">"radon"</span>, ylabel<span class="op">=</span><span class="st">"Density"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/tomas/Desktop/OSS/bambinos/bambi/bambi/models.py:845: FutureWarning: 'pps' has been replaced by 'response' and is not going to work in the future
  warnings.warn(
/tmp/ipykernel_45847/1213510270.py:9: FutureWarning: extract_dataset has been deprecated, please use extract
  pps = az.extract_dataset(pooled_results, group="posterior_predictive")["log_radon"].values</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-20-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The distributions look very similar, but we see that we have some extreme values. Hence if we need a number to compare them let us use the median.</p>
<div id="cell-47" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>np.median(a<span class="op">=</span>pps, axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<pre><code>array([3.862334  , 2.07121177])</code></pre>
</div>
</div>
<div id="cell-48" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>df.groupby([<span class="st">"floor"</span>])[<span class="st">"activity"</span>].median()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<pre><code>floor
Basement    3.9
Floor       2.1
Name: activity, dtype: float64</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="no-pooling" class="level3">
<h3 class="anchored" data-anchor-id="no-pooling">No Pooling</h3>
<p>The following model uses both <code>floor</code> and <code>county</code> as predictors. They are represented with an interaction effect. It means the predicted radon level for a given measurement depends both on the floor level as well as the county. This interaction coefficient allows the floor effect to vary across counties. Or said analogously, the county effect can vary across floor levels.</p>
<section id="model-1" class="level4">
<h4 class="anchored" data-anchor-id="model-1">Model</h4>
<p><span class="math display">\[
y = \gamma_{jk} + \varepsilon
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
y &amp;= \text{Response for the (log) radon measurement }\\
\gamma_{jk} &amp;= \text{Coefficient for floor level } j \text{ and county } k\\
\varepsilon &amp; = \text{Residual random error}
\end{aligned}
\]</span></p>
</section>
<section id="priors-1" class="level4">
<h4 class="anchored" data-anchor-id="priors-1">Priors</h4>
<section id="common-effects-1" class="level5">
<h5 class="anchored" data-anchor-id="common-effects-1">Common effects</h5>
<p>The common effect is the interaction between <code>floor</code> and <code>county</code>. The prior is</p>
<p><span class="math display">\[
\gamma_{jk} \sim \text{Normal}(0, \sigma_{\gamma_{jk}})
\]</span></p>
<p>for all <span class="math inline">\(j: 1, 2\)</span> and <span class="math inline">\(k: 1, \cdots, 85\)</span>.<br>
<span class="math inline">\(\sigma_{\gamma_{jk}}\)</span> is a positive constant that we set to 10 in all cases.</p>
</section>
<section id="residual-error-1" class="level5">
<h5 class="anchored" data-anchor-id="residual-error-1">Residual error</h5>
<p><span class="math display">\[
\begin{aligned}
\varepsilon_i &amp; \sim \text{Normal}(0, \sigma) \\
\sigma &amp; \sim \text{Exponential}(\lambda)
\end{aligned}
\]</span> where <span class="math inline">\(\lambda\)</span> is a positive constant that we set to 1.</p>
<p>To specify this model in Bambi we can use the formula <code>log_radon ~ 0 + county:floor</code>. Again, we remove the global intercept with the <code>0</code> on the right hand side. <code>county:floor</code> specifies the multiplicative interaction between <code>county</code> and <code>floor</code>.</p>
<div id="cell-53" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>unpooled_priors <span class="op">=</span> {</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"county:floor"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sigma"</span>: bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>unpooled_model <span class="op">=</span> bmb.Model(<span class="st">"log_radon ~ 0 + county:floor"</span>, df, priors<span class="op">=</span>unpooled_priors)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>unpooled_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<pre><code>       Formula: log_radon ~ 0 + county:floor
        Family: gaussian
          Link: mu = identity
  Observations: 919
        Priors: 
    target = mu
        Common-level effects
            county:floor ~ Normal(mu: 0.0, sigma: 10.0)
        
        Auxiliary parameters
            sigma ~ Exponential(lam: 1.0)</code></pre>
</div>
</div>
<div id="cell-54" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>unpooled_results <span class="op">=</span> unpooled_model.fit()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [sigma, county:floor]</code></pre>
</div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"afd44092af0c488fb8a5c3d5fc608bf2","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sampling 2 chains for 1_000 tune and 1_000 draw iterations (2_000 + 2_000 draws total) took 36 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details</code></pre>
</div>
</div>
<div id="cell-55" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>unpooled_model.graph()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-25-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>From the graph representation of the model we see the model estimates <span class="math inline">\(170 = 85 \times 2\)</span> parameters for the <code>county:floor</code> interaction. Let us now explore the model fit.</p>
<p>First, we can now see the plot of the marginal posterior distributions along with the sampling traces.</p>
<div id="cell-58" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>az.plot_trace(data<span class="op">=</span>unpooled_results, compact<span class="op">=</span><span class="va">True</span>, chain_prop<span class="op">=</span>{<span class="st">"ls"</span>: <span class="st">"-"</span>})</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Un-Pooled Model Trace"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-26-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Some posteriors for <code>county:floor</code> are much more spread than others, which makes it harder to compare them. To obtain a better summary visualization we can use a forest plot. This plot also allows us to identify exactly the combination of <code>county</code> and <code>floor</code> level.</p>
<div id="cell-60" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>az.plot_forest(data<span class="op">=</span>unpooled_results, figsize<span class="op">=</span>(<span class="dv">6</span>, <span class="dv">32</span>), r_hat<span class="op">=</span><span class="va">True</span>, combined<span class="op">=</span><span class="va">True</span>, textsize<span class="op">=</span><span class="dv">8</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-27-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Note how for the combination <code>county == 'YELLOW MEDICINE' and floor == 'Floor'</code> where we do not have any observations, the model can still generate predictions which are essentially coming from the prior distributions, which explains the large HDI intervals.</p>
<p>Next, let’s have a look into the posterior mean for each <code>county</code> and <code>floor</code> combination:</p>
<div id="cell-63" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>unpooled_summary <span class="op">=</span> az.summary(data<span class="op">=</span>unpooled_results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now plot the posterior distribution mean of the gamma coefficients against the observed values (sample).</p>
<div id="cell-65" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get county and floor names from summary table</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>var_mapping <span class="op">=</span> (</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>    unpooled_summary</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>    .iloc[:<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    .reset_index(drop<span class="op">=</span><span class="va">False</span>)[<span class="st">"index"</span>].<span class="bu">str</span>.<span class="bu">slice</span>(<span class="dv">13</span>, <span class="op">-</span><span class="dv">1</span>).<span class="bu">str</span>.split(<span class="st">","</span>).<span class="bu">apply</span>(pd.Series)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>var_mapping.rename(columns<span class="op">=</span>{<span class="dv">0</span>: <span class="st">"county"</span>, <span class="dv">1</span>: <span class="st">"floor"</span>}, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>var_mapping[<span class="st">"county"</span>] <span class="op">=</span> var_mapping[<span class="st">"county"</span>].<span class="bu">str</span>.strip()</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>var_mapping[<span class="st">"floor"</span>] <span class="op">=</span> var_mapping[<span class="st">"floor"</span>].<span class="bu">str</span>.strip()</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>var_mapping.index <span class="op">=</span> unpooled_summary.iloc[:<span class="op">-</span><span class="dv">1</span>].index</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge with observed values</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>unpooled_summary_2 <span class="op">=</span> pd.concat([var_mapping, unpooled_summary.iloc[:<span class="op">-</span><span class="dv">1</span>]], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">7</span>, <span class="dv">6</span>))</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>    unpooled_summary_2</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>    .merge(right<span class="op">=</span>log_radon_county_agg, on<span class="op">=</span>[<span class="st">"county"</span>, <span class="st">"floor"</span>], how<span class="op">=</span><span class="st">"left"</span>)</span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>    .pipe(</span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>        (sns.scatterplot, <span class="st">"data"</span>),</span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a>        x<span class="op">=</span><span class="st">"log_radon_mean"</span>,</span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>        y<span class="op">=</span><span class="st">"mean"</span>,</span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a>        hue<span class="op">=</span><span class="st">"floor"</span>,</span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a>        ax<span class="op">=</span>ax</span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a>ax.axline(xy1<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, label<span class="op">=</span><span class="st">"diagonal"</span>)</span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a>ax.legend()</span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(</span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"log(Radon + 0.1) Mean per County (Unpooled Model)"</span>,</span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span><span class="st">"observed (sample)"</span>,</span>
<span id="cb41-34"><a href="#cb41-34" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span><span class="st">"prediction"</span>,</span>
<span id="cb41-35"><a href="#cb41-35" aria-hidden="true" tabindex="-1"></a>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-29-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As expected, the values strongly concentrated along the diagonal. In other words, for each county and floor level combination, the model uses their sample mean of radon level as prediction, plus smoothing due to the non-flat priors.</p>
<p><strong>What does <em>no pooling</em> exactly mean here?</strong></p>
<p>In the previous example we said <em>complete pooling</em> means the observations are pooled together into single estimates no matter the county they belong to. The situation is completely the opposite in this <em>no pooling</em> scenario. Here, none of the measurements in a given county affect the computation of the coefficient for another county. That’s why, in the end, the estimation for each combination of county and floor level (i.e.&nbsp;<span class="math inline">\(\gamma_{jk}\)</span>) is the mean of the measurements in that county and floor level (plus prior smoothing) as is reflected in the diagonal scatterplot above.</p>
</section>
</section>
</section>
</section>
<section id="multilevel-and-hierarchical-models" class="level2">
<h2 class="anchored" data-anchor-id="multilevel-and-hierarchical-models">Multilevel and hierarchical models</h2>
<p>In this section we are going to explore various types of hierarchical models. If you’re familiar with the PyMC way of using hierarchies, the Bambi way (borrowed from mixed effects models way) may be a bit unfamiliar in the beginning, but as we will see, the notation is very convenient. A good explanation is found in <a href="https://www.bayesrulesbook.com/chapter-16.html">Chapter 16 from Bayes Rules book</a>, specifically <a href="https://www.bayesrulesbook.com/chapter-16.html#another-way-to-think-about-it">section 16.3.2</a>. Moreover, you can also take a look into the <a href="https://bambinos.github.io/bambi/main/examples.html">Bambi examples section</a> where you can find other use cases.</p>
<section id="partial-pooling-model" class="level3">
<h3 class="anchored" data-anchor-id="partial-pooling-model">Partial pooling model</h3>
<p>We start with a model that considers a global intercept and varying intercepts for each county. The dispersion parameter of the prior for these varying intercepts is an <em>hyperprior</em> that is common to all the counties. As we are going to conclude later, this is what causes the <strong>partial pooling</strong> in the model estimates.</p>
<section id="model-2" class="level4">
<h4 class="anchored" data-anchor-id="model-2">Model</h4>
<p>Let us use greek letters for common effects and roman letters for varying effects. In this case, <span class="math inline">\(\alpha\)</span> is a common intercept and <span class="math inline">\(u\)</span> is a group-specific intercept.</p>
<p><span class="math display">\[
y = \alpha + u_j + \varepsilon
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
y &amp;= \text{Response for the (log) radon measurement } \\
\alpha &amp;= \text{Intercept common to all measurements or global intercept} \\
u_j &amp;= \text{Intercept specific to the county } j \\
\varepsilon &amp; = \text{Residual random error}
\end{aligned}
\]</span></p>
</section>
<section id="priors-2" class="level4">
<h4 class="anchored" data-anchor-id="priors-2">Priors</h4>
<section id="common-effects-2" class="level5">
<h5 class="anchored" data-anchor-id="common-effects-2">Common effects</h5>
<p>The only common effect in this model is the intercept <span class="math inline">\(\alpha\)</span>. We have</p>
<p><span class="math display">\[
\alpha \sim \text{Normal}(0, \sigma_\alpha)
\]</span></p>
<p>where <span class="math inline">\(\sigma_\alpha\)</span> is a positive constant that we set to 10.</p>
</section>
<section id="group-specific-effects" class="level5">
<h5 class="anchored" data-anchor-id="group-specific-effects">Group-specific effects</h5>
<p><span class="math display">\[
u_j \sim \text{Normal}(0, \sigma_u)
\]</span></p>
<p>for all <span class="math inline">\(j: 1, \cdots, 85\)</span>.</p>
<p>Contrary to the common effects case, <span class="math inline">\(\sigma_u\)</span> is considered a random variable.</p>
<p>We assign <span class="math inline">\(\sigma_u\)</span> the following hyperprior, which is the same to all the counties,</p>
<p><span class="math display">\[
\sigma_u\sim \text{Exponential}(\tau)
\]</span></p>
<p>and <span class="math inline">\(\tau\)</span> is a positive constant that we set to <span class="math inline">\(1\)</span>.</p>
</section>
<section id="residual-error-2" class="level5">
<h5 class="anchored" data-anchor-id="residual-error-2">Residual error</h5>
<p><span class="math display">\[
\begin{aligned}
\varepsilon &amp; \sim \text{Normal}(0, \sigma) \\
\sigma &amp; \sim \text{Exponential}(\lambda)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\lambda\)</span> is a positive constant that we set to 1.</p>
</section>
</section>
<section id="notes" class="level4">
<h4 class="anchored" data-anchor-id="notes">Notes</h4>
<p>The common intercept <span class="math inline">\(\alpha\)</span> represents the mean response across all counties and floor levels.</p>
<p>On top of it, the county-specific intercept terms <span class="math inline">\(u_j\)</span> represent county-specific deviations from that global mean. This type of term is also known as a vaying intercept in the statistical literature.</p>
<div id="cell-73" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can add the hyper-priors inside the prior dictionary parameter of the model constructor</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>partial_pooling_priors <span class="op">=</span> {</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Intercept"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1|county"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span>bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>)),</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sigma"</span>: bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>partial_pooling_model <span class="op">=</span> bmb.Model(</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    formula<span class="op">=</span><span class="st">"log_radon ~ 1 + (1|county)"</span>, </span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df, </span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    priors<span class="op">=</span>partial_pooling_priors, </span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>    noncentered<span class="op">=</span><span class="va">False</span></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>partial_pooling_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="29">
<pre><code>       Formula: log_radon ~ 1 + (1|county)
        Family: gaussian
          Link: mu = identity
  Observations: 919
        Priors: 
    target = mu
        Common-level effects
            Intercept ~ Normal(mu: 0.0, sigma: 10.0)
        
        Group-level effects
            1|county ~ Normal(mu: 0.0, sigma: Exponential(lam: 1.0))
        
        Auxiliary parameters
            sigma ~ Exponential(lam: 1.0)</code></pre>
</div>
</div>
<p>The <code>noncentered</code> argument asks Bambi not to use the non centered representation for the varying effects. This makes the graph representation clearer and is closer to the original implementation in the PyMC documentation.</p>
<div id="cell-75" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>partial_pooling_results <span class="op">=</span> partial_pooling_model.fit()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [sigma, Intercept, 1|county_sigma, 1|county]</code></pre>
</div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"6828a1f20a024191af101b0169427205","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sampling 2 chains for 1_000 tune and 1_000 draw iterations (2_000 + 2_000 draws total) took 6 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details</code></pre>
</div>
</div>
<p>We can inspect the graphical representation of the model:</p>
<div id="cell-77" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>partial_pooling_model.graph()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-32-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can clearly see a new hierarchical level as compared to the complete pooling model and unpooled model.</p>
<p>Next, we can plot the posterior distribution of the coefficients in the model:</p>
<div id="cell-80" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>az.plot_trace(data<span class="op">=</span>partial_pooling_results, compact<span class="op">=</span><span class="va">True</span>, chain_prop<span class="op">=</span>{<span class="st">"ls"</span>: <span class="st">"-"</span>})</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Partial Pooling Model Trace"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-33-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><code>1|county</code> is <span class="math inline">\(u_j\)</span>, the county-specific intercepts.</li>
<li><code>1|county_sigma</code> is <span class="math inline">\(\sigma_u\)</span>, the standard deviation of the county-specific intercepts above.</li>
</ul>
<p>Let us now compare the posterior predictive mean against the observed data at county level.</p>
<div id="cell-83" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>partial_pooling_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="33">

            <div>
              <div class="xr-header">
                <div class="xr-obj-type">arviz.InferenceData</div>
              </div>
              <ul class="xr-sections group-sections">
              
            <li class="xr-section-item">
                  <input id="idata_posteriorbd377d2a-fcb9-48ff-9778-a75b7a3af237" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_posteriorbd377d2a-fcb9-48ff-9778-a75b7a3af237" class="xr-section-summary">posterior</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewbox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewbox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class="xr-text-repr-fallback">&lt;xarray.Dataset&gt; Size: 1MB
Dimensions:             (chain: 2, draw: 1000, county__factor_dim: 85)
Coordinates:
  * chain               (chain) int64 16B 0 1
  * draw                (draw) int64 8kB 0 1 2 3 4 5 ... 994 995 996 997 998 999
  * county__factor_dim  (county__factor_dim) &lt;U17 6kB 'AITKIN' ... 'YELLOW ME...
Data variables:
    1|county            (chain, draw, county__factor_dim) float64 1MB 0.1093 ...
    1|county_sigma      (chain, draw) float64 16kB 0.3421 0.3602 ... 0.2518
    Intercept           (chain, draw) float64 16kB 1.44 1.4 ... 1.301 1.295
    sigma               (chain, draw) float64 16kB 0.7651 0.7419 ... 0.7995
Attributes:
    created_at:                  2024-05-26T22:50:19.814875+00:00
    arviz_version:               0.18.0
    inference_library:           pymc
    inference_library_version:   5.15.0+23.g19be124e
    sampling_time:               6.147133111953735
    tuning_steps:                1000
    modeling_interface:          bambi
    modeling_interface_version:  0.13.1.dev39+gb7d6a6cb</pre><div class="xr-wrap" style="display:none"><div class="xr-header"><div class="xr-obj-type">xarray.Dataset</div></div><ul class="xr-sections"><li class="xr-section-item"><input id="section-59dff987-22f0-4f00-b52c-9db2963fc79f" class="xr-section-summary-in" type="checkbox" disabled=""><label for="section-59dff987-22f0-4f00-b52c-9db2963fc79f" class="xr-section-summary" title="Expand/collapse section">Dimensions:</label><div class="xr-section-inline-details"><ul class="xr-dim-list"><li><span class="xr-has-index">chain</span>: 2</li><li><span class="xr-has-index">draw</span>: 1000</li><li><span class="xr-has-index">county__factor_dim</span>: 85</li></ul></div><div class="xr-section-details"></div></li><li class="xr-section-item"><input id="section-aea6300a-95c2-414d-a9e8-c2f248cfa990" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-aea6300a-95c2-414d-a9e8-c2f248cfa990" class="xr-section-summary">Coordinates: <span>(3)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-var-name"><span class="xr-has-index">chain</span></div><div class="xr-var-dims">(chain)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">0 1</div><input id="attrs-6ffc359d-fa8f-4bea-9e9b-66328d2ac696" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-6ffc359d-fa8f-4bea-9e9b-66328d2ac696" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-e8be05cd-24d8-4b0c-8b5f-2fd3533e96b4" class="xr-var-data-in" type="checkbox"><label for="data-e8be05cd-24d8-4b0c-8b5f-2fd3533e96b4" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([0, 1])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span class="xr-has-index">draw</span></div><div class="xr-var-dims">(draw)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">0 1 2 3 4 5 ... 995 996 997 998 999</div><input id="attrs-9c5f3806-a223-4f88-9c7e-aa5dfda734fa" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-9c5f3806-a223-4f88-9c7e-aa5dfda734fa" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-c61fe115-b45f-491e-907e-ec968a5c1f75" class="xr-var-data-in" type="checkbox"><label for="data-c61fe115-b45f-491e-907e-ec968a5c1f75" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([  0,   1,   2, ..., 997, 998, 999])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span class="xr-has-index">county__factor_dim</span></div><div class="xr-var-dims">(county__factor_dim)</div><div class="xr-var-dtype">&lt;U17</div><div class="xr-var-preview xr-preview">'AITKIN' ... 'YELLOW MEDICINE'</div><input id="attrs-e276fcec-dc8f-4f1e-b2dd-71858b8e5188" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-e276fcec-dc8f-4f1e-b2dd-71858b8e5188" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-ba5672df-cd7f-44bd-b153-7ed53a5683b6" class="xr-var-data-in" type="checkbox"><label for="data-ba5672df-cd7f-44bd-b153-7ed53a5683b6" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array(['AITKIN', 'ANOKA', 'BECKER', 'BELTRAMI', 'BENTON', 'BIG STONE',
       'BLUE EARTH', 'BROWN', 'CARLTON', 'CARVER', 'CASS', 'CHIPPEWA',
       'CHISAGO', 'CLAY', 'CLEARWATER', 'COOK', 'COTTONWOOD', 'CROW WING',
       'DAKOTA', 'DODGE', 'DOUGLAS', 'FARIBAULT', 'FILLMORE', 'FREEBORN',
       'GOODHUE', 'HENNEPIN', 'HOUSTON', 'HUBBARD', 'ISANTI', 'ITASCA',
       'JACKSON', 'KANABEC', 'KANDIYOHI', 'KITTSON', 'KOOCHICHING',
       'LAC QUI PARLE', 'LAKE', 'LAKE OF THE WOODS', 'LE SUEUR', 'LINCOLN',
       'LYON', 'MAHNOMEN', 'MARSHALL', 'MARTIN', 'MCLEOD', 'MEEKER',
       'MILLE LACS', 'MORRISON', 'MOWER', 'MURRAY', 'NICOLLET', 'NOBLES',
       'NORMAN', 'OLMSTED', 'OTTER TAIL', 'PENNINGTON', 'PINE', 'PIPESTONE',
       'POLK', 'POPE', 'RAMSEY', 'REDWOOD', 'RENVILLE', 'RICE', 'ROCK',
       'ROSEAU', 'SCOTT', 'SHERBURNE', 'SIBLEY', 'ST LOUIS', 'STEARNS',
       'STEELE', 'STEVENS', 'SWIFT', 'TODD', 'TRAVERSE', 'WABASHA', 'WADENA',
       'WASECA', 'WASHINGTON', 'WATONWAN', 'WILKIN', 'WINONA', 'WRIGHT',
       'YELLOW MEDICINE'], dtype='&lt;U17')</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-a88c07d5-d489-4e86-9a34-b63533bb296b" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-a88c07d5-d489-4e86-9a34-b63533bb296b" class="xr-section-summary">Data variables: <span>(4)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-var-name"><span>1|county</span></div><div class="xr-var-dims">(chain, draw, county__factor_dim)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.1093 -0.5662 ... 0.325 -0.2251</div><input id="attrs-082fcb2e-3102-4eaa-b18d-186ea861abc8" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-082fcb2e-3102-4eaa-b18d-186ea861abc8" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-f7048c73-e12e-4a98-b5d7-78f6961404dd" class="xr-var-data-in" type="checkbox"><label for="data-f7048c73-e12e-4a98-b5d7-78f6961404dd" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[[ 0.10925731, -0.566226  , -0.09000699, ..., -0.27306477,
          0.23927698,  0.20538249],
        [-0.61860104, -0.54807003, -0.17833427, ..., -0.13802683,
         -0.29767395, -0.05617042],
        [-0.02315516, -0.63490798, -0.1897692 , ..., -0.06686441,
          0.14023241,  0.02202231],
        ...,
        [-0.30931644, -0.46458741, -0.37153588, ...,  0.11928815,
          0.10271827,  0.12858761],
        [-0.24246309, -0.42452553,  0.10190816, ...,  0.0297322 ,
          0.31775183,  0.12238267],
        [-0.11289643, -0.45281249, -0.0720627 , ..., -0.00316119,
          0.109228  , -0.15858485]],

       [[-0.48851576, -0.45108672,  0.09534777, ...,  0.40064506,
          0.29706353, -0.5736238 ],
        [-0.38335947, -0.45444664, -0.04944886, ...,  0.32466663,
          0.46604097, -0.09875201],
        [-0.172785  , -0.31658689, -0.02637731, ...,  0.02700598,
          0.03342428,  0.17974528],
        ...,
        [ 0.10934748, -0.39886058,  0.27313887, ...,  0.16720247,
          0.04149402, -0.07026582],
        [-0.22966455, -0.24279823, -0.41640239, ...,  0.00669983,
          0.16205345, -0.23887474],
        [-0.06651367, -0.42916719,  0.15400986, ...,  0.34961275,
          0.32496288, -0.22505465]]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>1|county_sigma</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.3421 0.3602 ... 0.2498 0.2518</div><input id="attrs-7c535837-ab45-4934-8a60-7e7cca314f6b" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-7c535837-ab45-4934-8a60-7e7cca314f6b" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-a5a98d4f-87b0-4e38-ad2b-a2b17d44cad2" class="xr-var-data-in" type="checkbox"><label for="data-a5a98d4f-87b0-4e38-ad2b-a2b17d44cad2" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.34210321, 0.36021532, 0.29334103, ..., 0.28282012, 0.35124886,
        0.33627169],
       [0.31875038, 0.35045077, 0.33136916, ..., 0.22967175, 0.24982968,
        0.25178762]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>Intercept</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">1.44 1.4 1.372 ... 1.301 1.295</div><input id="attrs-1484de35-3884-4148-baea-7791e38488c6" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-1484de35-3884-4148-baea-7791e38488c6" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-721cbb8a-557d-4276-97ee-614b8c093a0d" class="xr-var-data-in" type="checkbox"><label for="data-721cbb8a-557d-4276-97ee-614b8c093a0d" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[1.44017201, 1.40006556, 1.37197719, ..., 1.37240347, 1.31568418,
        1.36887556],
       [1.34806946, 1.31749286, 1.31428794, ..., 1.33580165, 1.30139677,
        1.29451323]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>sigma</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.7651 0.7419 ... 0.7604 0.7995</div><input id="attrs-f64eecea-b1eb-43c3-a92f-a92c8128ada0" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-f64eecea-b1eb-43c3-a92f-a92c8128ada0" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-bca69a3f-0467-41d6-b543-27194a7b79b8" class="xr-var-data-in" type="checkbox"><label for="data-bca69a3f-0467-41d6-b543-27194a7b79b8" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.76509665, 0.74185114, 0.80634063, ..., 0.76435342, 0.76195412,
        0.7711739 ],
       [0.75649317, 0.73755021, 0.7870685 , ..., 0.75311483, 0.76044651,
        0.79951296]])</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-151569cd-17f2-4171-afc7-1ddb87e727a4" class="xr-section-summary-in" type="checkbox"><label for="section-151569cd-17f2-4171-afc7-1ddb87e727a4" class="xr-section-summary">Indexes: <span>(3)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-index-name"><div>chain</div></div><div class="xr-index-preview">PandasIndex</div><div></div><input id="index-0cdb93c5-d836-4ffd-a23f-5fccd0a24e4d" class="xr-index-data-in" type="checkbox"><label for="index-0cdb93c5-d836-4ffd-a23f-5fccd0a24e4d" title="Show/Hide index repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-index-data"><pre>PandasIndex(Index([0, 1], dtype='int64', name='chain'))</pre></div></li><li class="xr-var-item"><div class="xr-index-name"><div>draw</div></div><div class="xr-index-preview">PandasIndex</div><div></div><input id="index-94d742d1-0183-4051-83c4-d134084b5011" class="xr-index-data-in" type="checkbox"><label for="index-94d742d1-0183-4051-83c4-d134084b5011" title="Show/Hide index repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-index-data"><pre>PandasIndex(Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
       ...
       990, 991, 992, 993, 994, 995, 996, 997, 998, 999],
      dtype='int64', name='draw', length=1000))</pre></div></li><li class="xr-var-item"><div class="xr-index-name"><div>county__factor_dim</div></div><div class="xr-index-preview">PandasIndex</div><div></div><input id="index-b7b3c355-4c83-4473-8099-51508808e41e" class="xr-index-data-in" type="checkbox"><label for="index-b7b3c355-4c83-4473-8099-51508808e41e" title="Show/Hide index repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-index-data"><pre>PandasIndex(Index(['AITKIN', 'ANOKA', 'BECKER', 'BELTRAMI', 'BENTON', 'BIG STONE',
       'BLUE EARTH', 'BROWN', 'CARLTON', 'CARVER', 'CASS', 'CHIPPEWA',
       'CHISAGO', 'CLAY', 'CLEARWATER', 'COOK', 'COTTONWOOD', 'CROW WING',
       'DAKOTA', 'DODGE', 'DOUGLAS', 'FARIBAULT', 'FILLMORE', 'FREEBORN',
       'GOODHUE', 'HENNEPIN', 'HOUSTON', 'HUBBARD', 'ISANTI', 'ITASCA',
       'JACKSON', 'KANABEC', 'KANDIYOHI', 'KITTSON', 'KOOCHICHING',
       'LAC QUI PARLE', 'LAKE', 'LAKE OF THE WOODS', 'LE SUEUR', 'LINCOLN',
       'LYON', 'MAHNOMEN', 'MARSHALL', 'MARTIN', 'MCLEOD', 'MEEKER',
       'MILLE LACS', 'MORRISON', 'MOWER', 'MURRAY', 'NICOLLET', 'NOBLES',
       'NORMAN', 'OLMSTED', 'OTTER TAIL', 'PENNINGTON', 'PINE', 'PIPESTONE',
       'POLK', 'POPE', 'RAMSEY', 'REDWOOD', 'RENVILLE', 'RICE', 'ROCK',
       'ROSEAU', 'SCOTT', 'SHERBURNE', 'SIBLEY', 'ST LOUIS', 'STEARNS',
       'STEELE', 'STEVENS', 'SWIFT', 'TODD', 'TRAVERSE', 'WABASHA', 'WADENA',
       'WASECA', 'WASHINGTON', 'WATONWAN', 'WILKIN', 'WINONA', 'WRIGHT',
       'YELLOW MEDICINE'],
      dtype='object', name='county__factor_dim'))</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-17ea2065-6c05-4e81-8396-9fd02460b902" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-17ea2065-6c05-4e81-8396-9fd02460b902" class="xr-section-summary">Attributes: <span>(8)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><dl class="xr-attrs"><dt><span>created_at :</span></dt><dd>2024-05-26T22:50:19.814875+00:00</dd><dt><span>arviz_version :</span></dt><dd>0.18.0</dd><dt><span>inference_library :</span></dt><dd>pymc</dd><dt><span>inference_library_version :</span></dt><dd>5.15.0+23.g19be124e</dd><dt><span>sampling_time :</span></dt><dd>6.147133111953735</dd><dt><span>tuning_steps :</span></dt><dd>1000</dd><dt><span>modeling_interface :</span></dt><dd>bambi</dd><dt><span>modeling_interface_version :</span></dt><dd>0.13.1.dev39+gb7d6a6cb</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>
            
            <li class="xr-section-item">
                  <input id="idata_sample_statsecc5ef60-e8d1-4260-b528-de4a11e45925" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_sample_statsecc5ef60-e8d1-4260-b528-de4a11e45925" class="xr-section-summary">sample_stats</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewbox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewbox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class="xr-text-repr-fallback">&lt;xarray.Dataset&gt; Size: 252kB
Dimensions:                (chain: 2, draw: 1000)
Coordinates:
  * chain                  (chain) int64 16B 0 1
  * draw                   (draw) int64 8kB 0 1 2 3 4 5 ... 995 996 997 998 999
Data variables: (12/17)
    acceptance_rate        (chain, draw) float64 16kB 0.9844 0.986 ... 0.8069
    diverging              (chain, draw) bool 2kB False False ... False False
    energy                 (chain, draw) float64 16kB 1.133e+03 ... 1.111e+03
    energy_error           (chain, draw) float64 16kB 0.1159 -0.4477 ... 0.4183
    index_in_trajectory    (chain, draw) int64 16kB 5 -4 4 -4 -2 ... 6 -2 -4 -4
    largest_eigval         (chain, draw) float64 16kB nan nan nan ... nan nan
    ...                     ...
    process_time_diff      (chain, draw) float64 16kB 0.001789 ... 0.001452
    reached_max_treedepth  (chain, draw) bool 2kB False False ... False False
    smallest_eigval        (chain, draw) float64 16kB nan nan nan ... nan nan
    step_size              (chain, draw) float64 16kB 0.3948 0.3948 ... 0.3688
    step_size_bar          (chain, draw) float64 16kB 0.467 0.467 ... 0.4521
    tree_depth             (chain, draw) int64 16kB 3 3 3 3 3 3 ... 3 3 3 3 3 3
Attributes:
    created_at:                  2024-05-26T22:50:19.833200+00:00
    arviz_version:               0.18.0
    inference_library:           pymc
    inference_library_version:   5.15.0+23.g19be124e
    sampling_time:               6.147133111953735
    tuning_steps:                1000
    modeling_interface:          bambi
    modeling_interface_version:  0.13.1.dev39+gb7d6a6cb</pre><div class="xr-wrap" style="display:none"><div class="xr-header"><div class="xr-obj-type">xarray.Dataset</div></div><ul class="xr-sections"><li class="xr-section-item"><input id="section-620a958a-c276-4b77-8b4b-a793d76addeb" class="xr-section-summary-in" type="checkbox" disabled=""><label for="section-620a958a-c276-4b77-8b4b-a793d76addeb" class="xr-section-summary" title="Expand/collapse section">Dimensions:</label><div class="xr-section-inline-details"><ul class="xr-dim-list"><li><span class="xr-has-index">chain</span>: 2</li><li><span class="xr-has-index">draw</span>: 1000</li></ul></div><div class="xr-section-details"></div></li><li class="xr-section-item"><input id="section-6b9e3fed-c1f3-45f0-b73f-68d1a50da0ce" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-6b9e3fed-c1f3-45f0-b73f-68d1a50da0ce" class="xr-section-summary">Coordinates: <span>(2)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-var-name"><span class="xr-has-index">chain</span></div><div class="xr-var-dims">(chain)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">0 1</div><input id="attrs-7d67c0e0-8fe8-4834-b26b-d642461606cf" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-7d67c0e0-8fe8-4834-b26b-d642461606cf" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-e1a7c9b0-9019-400a-b890-678c063a94f3" class="xr-var-data-in" type="checkbox"><label for="data-e1a7c9b0-9019-400a-b890-678c063a94f3" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([0, 1])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span class="xr-has-index">draw</span></div><div class="xr-var-dims">(draw)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">0 1 2 3 4 5 ... 995 996 997 998 999</div><input id="attrs-b39499b8-fe85-4813-ad96-79c420ea5387" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-b39499b8-fe85-4813-ad96-79c420ea5387" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-7bbcb5da-d077-4a4d-a184-9774eea70b3c" class="xr-var-data-in" type="checkbox"><label for="data-7bbcb5da-d077-4a4d-a184-9774eea70b3c" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([  0,   1,   2, ..., 997, 998, 999])</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-72fc4c38-8fbd-4db1-99e7-054c93404c7d" class="xr-section-summary-in" type="checkbox"><label for="section-72fc4c38-8fbd-4db1-99e7-054c93404c7d" class="xr-section-summary">Data variables: <span>(17)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-var-name"><span>acceptance_rate</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.9844 0.986 ... 0.9234 0.8069</div><input id="attrs-7494cb4a-fe5a-4f45-8bbf-870ba2b94d68" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-7494cb4a-fe5a-4f45-8bbf-870ba2b94d68" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-8b546e9b-290f-4936-8dc9-947b4d321ec8" class="xr-var-data-in" type="checkbox"><label for="data-8b546e9b-290f-4936-8dc9-947b4d321ec8" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.98436784, 0.98601233, 0.96124626, ..., 0.22486349, 0.33767736,
        0.63248875],
       [0.97609702, 0.54231738, 0.66807577, ..., 0.90569141, 0.92340586,
        0.80685305]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>diverging</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">bool</div><div class="xr-var-preview xr-preview">False False False ... False False</div><input id="attrs-abec51d8-5e62-4232-93fd-3d7950793702" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-abec51d8-5e62-4232-93fd-3d7950793702" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-5def41b4-41b9-4117-ae1e-9fd73ba12813" class="xr-var-data-in" type="checkbox"><label for="data-5def41b4-41b9-4117-ae1e-9fd73ba12813" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>energy</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">1.133e+03 1.142e+03 ... 1.111e+03</div><input id="attrs-32d9a1ab-1c67-4444-88eb-828218d7eabc" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-32d9a1ab-1c67-4444-88eb-828218d7eabc" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-6e728650-4277-4825-bfc2-48d432782f66" class="xr-var-data-in" type="checkbox"><label for="data-6e728650-4277-4825-bfc2-48d432782f66" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[1132.54289023, 1141.58447928, 1132.55888828, ..., 1124.70123514,
        1120.42203721, 1137.91003849],
       [1126.17947876, 1137.18871967, 1139.57160506, ..., 1132.34943256,
        1117.88735552, 1110.58936884]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>energy_error</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.1159 -0.4477 ... -0.3531 0.4183</div><input id="attrs-d02c28fb-3f81-4d78-b4c2-e42720a31ac3" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-d02c28fb-3f81-4d78-b4c2-e42720a31ac3" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-972b164b-f926-4196-974e-a26651269b63" class="xr-var-data-in" type="checkbox"><label for="data-972b164b-f926-4196-974e-a26651269b63" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[ 0.11588806, -0.44768874, -0.21160451, ...,  0.        ,
         1.03675117,  0.92922675],
       [-0.57963706,  0.26236369,  0.52248315, ..., -0.62318029,
        -0.3531464 ,  0.41828798]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>index_in_trajectory</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">5 -4 4 -4 -2 5 ... -5 6 6 -2 -4 -4</div><input id="attrs-dfba7d24-5cc4-44a1-a39b-332beea9f1ac" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-dfba7d24-5cc4-44a1-a39b-332beea9f1ac" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-ff1d3786-d395-481a-8b51-3dd69d1243e0" class="xr-var-data-in" type="checkbox"><label for="data-ff1d3786-d395-481a-8b51-3dd69d1243e0" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[ 5, -4,  4, ...,  0,  6, -3],
       [-4, -2,  6, ..., -2, -4, -4]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>largest_eigval</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">nan nan nan nan ... nan nan nan nan</div><input id="attrs-80bea030-6c1a-441d-ba47-87c84ec4952c" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-80bea030-6c1a-441d-ba47-87c84ec4952c" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-4866264b-727d-4418-9057-fdc717767149" class="xr-var-data-in" type="checkbox"><label for="data-4866264b-727d-4418-9057-fdc717767149" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>lp</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">-1.101e+03 ... -1.078e+03</div><input id="attrs-5700b14c-a7af-416b-b9a6-6a4fe1c85c64" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-5700b14c-a7af-416b-b9a6-6a4fe1c85c64" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-a74d5af0-7821-4174-ba40-6abfb1d1a8e6" class="xr-var-data-in" type="checkbox"><label for="data-a74d5af0-7821-4174-ba40-6abfb1d1a8e6" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[-1100.61605617, -1089.20456285, -1085.37306441, ...,
        -1070.8425698 , -1081.67979406, -1099.51643907],
       [-1083.87244587, -1091.09546455, -1098.29189477, ...,
        -1078.38308382, -1070.94143721, -1077.73397692]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>max_energy_error</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">-0.8364 -0.4477 ... 0.4791 0.4183</div><input id="attrs-3a31b396-08e8-4640-9c1b-2026d9159e3a" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-3a31b396-08e8-4640-9c1b-2026d9159e3a" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-7ece8bad-b858-4237-815d-a03e157cc3a6" class="xr-var-data-in" type="checkbox"><label for="data-7ece8bad-b858-4237-815d-a03e157cc3a6" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[-0.83641232, -0.44768874, -0.21160451, ...,  2.57007364,
         1.75862866,  1.02220945],
       [-0.90128732,  1.19021079,  0.93406381, ..., -0.94764139,
         0.47910821,  0.41828798]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>n_steps</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">7.0 7.0 7.0 7.0 ... 7.0 7.0 7.0 7.0</div><input id="attrs-737438a5-f4e0-4546-bee0-378329eaa820" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-737438a5-f4e0-4546-bee0-378329eaa820" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-c0ecb801-2fbb-423a-87ac-5c0217823db7" class="xr-var-data-in" type="checkbox"><label for="data-c0ecb801-2fbb-423a-87ac-5c0217823db7" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[7., 7., 7., ..., 7., 7., 7.],
       [7., 7., 7., ..., 7., 7., 7.]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>perf_counter_diff</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.001788 0.002915 ... 0.001451</div><input id="attrs-05704835-0237-4e87-aabd-183ffeda0c5a" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-05704835-0237-4e87-aabd-183ffeda0c5a" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-abcbe03c-ad87-40e2-9f3f-2c37f29c3498" class="xr-var-data-in" type="checkbox"><label for="data-abcbe03c-ad87-40e2-9f3f-2c37f29c3498" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.00178769, 0.00291476, 0.00231155, ..., 0.0017955 , 0.00293924,
        0.00292779],
       [0.00228063, 0.00151575, 0.00228206, ..., 0.00143171, 0.00142564,
        0.00145092]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>perf_counter_start</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">1.254e+04 1.254e+04 ... 1.254e+04</div><input id="attrs-135fa754-b0ab-4ad9-909f-507f25775ef8" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-135fa754-b0ab-4ad9-909f-507f25775ef8" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-5a635329-a231-467b-b64c-f16cec8dbafe" class="xr-var-data-in" type="checkbox"><label for="data-5a635329-a231-467b-b64c-f16cec8dbafe" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[12537.35090269, 12537.35459767, 12537.35781169, ...,
        12539.52664598, 12539.52994777, 12539.53319831],
       [12537.46978133, 12537.47228786, 12537.47398005, ...,
        12539.76039464, 12539.76199041, 12539.76357576]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>process_time_diff</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.001789 0.002916 ... 0.001452</div><input id="attrs-25443c2f-4083-4d02-ab81-606f85f90afb" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-25443c2f-4083-4d02-ab81-606f85f90afb" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-47b10054-f23a-470a-a2fb-5839095c5ee3" class="xr-var-data-in" type="checkbox"><label for="data-47b10054-f23a-470a-a2fb-5839095c5ee3" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.00178875, 0.00291601, 0.00231246, ..., 0.0017962 , 0.00294044,
        0.00292832],
       [0.00228178, 0.00151654, 0.00228368, ..., 0.00143229, 0.00142587,
        0.00145171]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>reached_max_treedepth</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">bool</div><div class="xr-var-preview xr-preview">False False False ... False False</div><input id="attrs-9bb61afb-6805-4151-b214-f530ba441aa2" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-9bb61afb-6805-4151-b214-f530ba441aa2" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-c92d04d8-2bd7-4750-98e0-2ade7df08644" class="xr-var-data-in" type="checkbox"><label for="data-c92d04d8-2bd7-4750-98e0-2ade7df08644" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>smallest_eigval</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">nan nan nan nan ... nan nan nan nan</div><input id="attrs-c4a22e7d-f082-473f-8ccb-4a960f829526" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-c4a22e7d-f082-473f-8ccb-4a960f829526" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-dd5ddd39-81e9-4165-bdbd-428b4e8985cc" class="xr-var-data-in" type="checkbox"><label for="data-dd5ddd39-81e9-4165-bdbd-428b4e8985cc" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>step_size</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.3948 0.3948 ... 0.3688 0.3688</div><input id="attrs-599a7c46-c5fa-40ee-85f5-ff1874bf6e94" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-599a7c46-c5fa-40ee-85f5-ff1874bf6e94" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-6226f934-28cf-4e30-8db9-7f5a1779edf5" class="xr-var-data-in" type="checkbox"><label for="data-6226f934-28cf-4e30-8db9-7f5a1779edf5" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.39478295, 0.39478295, 0.39478295, ..., 0.39478295, 0.39478295,
        0.39478295],
       [0.36879817, 0.36879817, 0.36879817, ..., 0.36879817, 0.36879817,
        0.36879817]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>step_size_bar</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">0.467 0.467 0.467 ... 0.4521 0.4521</div><input id="attrs-7d033bd7-9c04-4aed-bbff-af736ddf80d1" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-7d033bd7-9c04-4aed-bbff-af736ddf80d1" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-b5565714-2b4f-47a5-88ac-029b2f99f4d6" class="xr-var-data-in" type="checkbox"><label for="data-b5565714-2b4f-47a5-88ac-029b2f99f4d6" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[0.46704576, 0.46704576, 0.46704576, ..., 0.46704576, 0.46704576,
        0.46704576],
       [0.45208578, 0.45208578, 0.45208578, ..., 0.45208578, 0.45208578,
        0.45208578]])</pre></div></li><li class="xr-var-item"><div class="xr-var-name"><span>tree_depth</span></div><div class="xr-var-dims">(chain, draw)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">3 3 3 3 3 3 3 4 ... 3 3 3 3 3 3 3 3</div><input id="attrs-b2ccfe2c-15a3-4f89-a996-ac1d628be972" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-b2ccfe2c-15a3-4f89-a996-ac1d628be972" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-d842fc97-97ce-4dc1-826e-4bb5a0592f8d" class="xr-var-data-in" type="checkbox"><label for="data-d842fc97-97ce-4dc1-826e-4bb5a0592f8d" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([[3, 3, 3, ..., 3, 3, 3],
       [3, 3, 3, ..., 3, 3, 3]])</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-968ad3ff-087c-4cf3-a5a5-e08274a37414" class="xr-section-summary-in" type="checkbox"><label for="section-968ad3ff-087c-4cf3-a5a5-e08274a37414" class="xr-section-summary">Indexes: <span>(2)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-index-name"><div>chain</div></div><div class="xr-index-preview">PandasIndex</div><div></div><input id="index-7915e612-508b-477c-b047-856ab356b0c3" class="xr-index-data-in" type="checkbox"><label for="index-7915e612-508b-477c-b047-856ab356b0c3" title="Show/Hide index repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-index-data"><pre>PandasIndex(Index([0, 1], dtype='int64', name='chain'))</pre></div></li><li class="xr-var-item"><div class="xr-index-name"><div>draw</div></div><div class="xr-index-preview">PandasIndex</div><div></div><input id="index-fca0a25d-0058-4cc7-864d-3125400f9833" class="xr-index-data-in" type="checkbox"><label for="index-fca0a25d-0058-4cc7-864d-3125400f9833" title="Show/Hide index repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-index-data"><pre>PandasIndex(Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
       ...
       990, 991, 992, 993, 994, 995, 996, 997, 998, 999],
      dtype='int64', name='draw', length=1000))</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-74ee933f-11b8-458a-a150-8b906c5d61d2" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-74ee933f-11b8-458a-a150-8b906c5d61d2" class="xr-section-summary">Attributes: <span>(8)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><dl class="xr-attrs"><dt><span>created_at :</span></dt><dd>2024-05-26T22:50:19.833200+00:00</dd><dt><span>arviz_version :</span></dt><dd>0.18.0</dd><dt><span>inference_library :</span></dt><dd>pymc</dd><dt><span>inference_library_version :</span></dt><dd>5.15.0+23.g19be124e</dd><dt><span>sampling_time :</span></dt><dd>6.147133111953735</dd><dt><span>tuning_steps :</span></dt><dd>1000</dd><dt><span>modeling_interface :</span></dt><dd>bambi</dd><dt><span>modeling_interface_version :</span></dt><dd>0.13.1.dev39+gb7d6a6cb</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>
            
            <li class="xr-section-item">
                  <input id="idata_observed_data85a2ff82-818e-42d5-9f0d-5cef0a997d26" class="xr-section-summary-in" type="checkbox">
                  <label for="idata_observed_data85a2ff82-818e-42d5-9f0d-5cef0a997d26" class="xr-section-summary">observed_data</label>
                  <div class="xr-section-inline-details"></div>
                  <div class="xr-section-details">
                      <ul id="xr-dataset-coord-list" class="xr-var-list">
                          <div style="padding-left:2rem;"><div><svg style="position: absolute; width: 0; height: 0; overflow: hidden">
<defs>
<symbol id="icon-database" viewbox="0 0 32 32">
<path d="M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z"></path>
<path d="M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
<path d="M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z"></path>
</symbol>
<symbol id="icon-file-text2" viewbox="0 0 32 32">
<path d="M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z"></path>
<path d="M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
<path d="M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z"></path>
</symbol>
</defs>
</svg>
<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body[data-theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block !important;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-index-preview {
  grid-column: 2 / 5;
  color: var(--xr-font-color2);
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data,
.xr-index-data-in:checked ~ .xr-index-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-index-name div,
.xr-index-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data,
.xr-index-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2,
.xr-no-icon {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
</style><pre class="xr-text-repr-fallback">&lt;xarray.Dataset&gt; Size: 15kB
Dimensions:    (__obs__: 919)
Coordinates:
  * __obs__    (__obs__) int64 7kB 0 1 2 3 4 5 6 ... 912 913 914 915 916 917 918
Data variables:
    log_radon  (__obs__) float64 7kB 1.435 1.03 0.2624 ... 0.0 2.219 0.8329
Attributes:
    created_at:                  2024-05-26T22:50:19.839116+00:00
    arviz_version:               0.18.0
    inference_library:           pymc
    inference_library_version:   5.15.0+23.g19be124e
    modeling_interface:          bambi
    modeling_interface_version:  0.13.1.dev39+gb7d6a6cb</pre><div class="xr-wrap" style="display:none"><div class="xr-header"><div class="xr-obj-type">xarray.Dataset</div></div><ul class="xr-sections"><li class="xr-section-item"><input id="section-93427ce2-0f79-4d85-9adc-6bd1100268a2" class="xr-section-summary-in" type="checkbox" disabled=""><label for="section-93427ce2-0f79-4d85-9adc-6bd1100268a2" class="xr-section-summary" title="Expand/collapse section">Dimensions:</label><div class="xr-section-inline-details"><ul class="xr-dim-list"><li><span class="xr-has-index">__obs__</span>: 919</li></ul></div><div class="xr-section-details"></div></li><li class="xr-section-item"><input id="section-41a805e5-358e-425c-aba5-70ced510c2e9" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-41a805e5-358e-425c-aba5-70ced510c2e9" class="xr-section-summary">Coordinates: <span>(1)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-var-name"><span class="xr-has-index">__obs__</span></div><div class="xr-var-dims">(__obs__)</div><div class="xr-var-dtype">int64</div><div class="xr-var-preview xr-preview">0 1 2 3 4 5 ... 914 915 916 917 918</div><input id="attrs-4302371e-e306-4f7b-b577-3a9405c1b938" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-4302371e-e306-4f7b-b577-3a9405c1b938" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-dc4ab078-a0a6-4efb-b7ce-e50584018898" class="xr-var-data-in" type="checkbox"><label for="data-dc4ab078-a0a6-4efb-b7ce-e50584018898" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([  0,   1,   2, ..., 916, 917, 918])</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-df672160-e55c-4638-a155-55a27518264c" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-df672160-e55c-4638-a155-55a27518264c" class="xr-section-summary">Data variables: <span>(1)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-var-name"><span>log_radon</span></div><div class="xr-var-dims">(__obs__)</div><div class="xr-var-dtype">float64</div><div class="xr-var-preview xr-preview">1.435 1.03 0.2624 ... 2.219 0.8329</div><input id="attrs-f0d28335-aff3-4a3f-a1e0-0d6bc575929d" class="xr-var-attrs-in" type="checkbox" disabled=""><label for="attrs-f0d28335-aff3-4a3f-a1e0-0d6bc575929d" title="Show/Hide attributes"><svg class="icon xr-icon-file-text2"><use href="#icon-file-text2"></use></svg></label><input id="data-ccd48d59-fc8d-4566-a8ff-21bf2e220668" class="xr-var-data-in" type="checkbox"><label for="data-ccd48d59-fc8d-4566-a8ff-21bf2e220668" title="Show/Hide data repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-var-attrs"><dl class="xr-attrs"></dl></div><div class="xr-var-data"><pre>array([ 1.43508453,  1.02961942,  0.26236426,  1.28093385,  1.7227666 ,
        1.7227666 ,  0.26236426,  1.60943791,  1.41098697,  1.28093385,
        0.95551145,  0.26236426,  1.02961942,  0.58778666,  1.16315081,
       -0.22314355,  0.09531018,  0.69314718,  1.36097655,  2.19722458,
        2.01490302,  1.80828877,  1.66770682,  1.84054963,  2.16332303,
        1.45861502,  1.77495235,  1.43508453,  1.06471074,  0.69314718,
        0.26236426,  0.47000363,  2.2512918 ,  0.58778666,  2.50143595,
        1.94591015,  0.78845736,  2.27212589,  1.25276297,  1.93152141,
        1.30833282,  0.83290912,  0.99325177,  0.78845736,  1.96009478,
        0.26236426,  1.36097655,  1.28093385,  1.36097655,  2.28238239,
        1.87180218,  1.54756251,  1.19392247,  0.95551145,  1.06471074,
        1.16315081,  0.53062825,  1.56861592,  1.41098697,  1.62924054,
        0.47000363,  1.58923521,  0.87546874, -0.10536052,  0.87546874,
        1.54756251,  2.40694511,  2.7080502 ,  2.16332303,  1.5260563 ,
        0.47000363,  1.38629436,  0.64185389,  0.53062825,  0.91629073,
        1.36097655,  1.64865863,  1.70474809,  1.74046617,  2.94968834,
        1.13140211,  1.64865863,  2.05412373,  2.10413415,  1.56861592,
        2.14006616,  0.53062825,  2.44234704,  3.2308044 ,  2.34180581,
        1.30833282,  1.02961942,  1.41098697,  0.74193734,  2.44234704,
        2.3321439 ,  0.26236426,  1.19392247,  1.48160454,  0.83290912,
...
        0.40546511,  0.95551145,  1.06471074,  0.53062825,  1.06471074,
        0.95551145,  2.32238772,  2.54160199,  0.78845736,  1.13140211,
       -2.30258509,  1.06471074,  0.33647224,  2.43361336,  0.33647224,
        0.        ,  1.5260563 ,  1.48160454,  1.09861229,  1.45861502,
        1.28093385,  1.94591015,  0.47000363, -0.51082562,  0.        ,
        0.18232156,  0.        , -0.51082562,  1.33500107, -0.10536052,
        1.06471074,  0.83290912,  1.58923521,  0.18232156,  1.09861229,
        0.53062825,  3.23867845,  0.40546511,  2.69462718,  3.03495299,
        0.91629073,  0.58778666, -0.10536052,  0.58778666,  1.06471074,
        1.98787435,  1.91692261,  0.95551145,  0.09531018,  0.95551145,
        0.        , -2.30258509,  2.41591378,  1.19392247, -0.22314355,
        0.83290912,  1.58923521,  1.94591015,  0.18232156,  0.64185389,
        0.95551145,  1.28093385,  0.        ,  0.09531018,  0.99325177,
        0.47000363, -2.30258509,  0.        ,  1.77495235,  1.28093385,
        0.78845736,  2.29253476,  1.94591015,  1.74046617,  0.83290912,
        1.80828877,  0.18232156,  1.48160454,  1.30833282,  1.25276297,
        0.26236426,  0.58778666,  1.45861502, -0.10536052,  2.96527307,
        0.95551145,  0.78845736,  0.33647224,  0.74193734,  1.33500107,
       -0.51082562,  0.09531018,  0.40546511, -0.69314718, -0.51082562,
        0.53062825,  0.        ,  2.21920348,  0.83290912])</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-c7bdb885-70f4-46ad-8517-2d866cee4cfd" class="xr-section-summary-in" type="checkbox"><label for="section-c7bdb885-70f4-46ad-8517-2d866cee4cfd" class="xr-section-summary">Indexes: <span>(1)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><ul class="xr-var-list"><li class="xr-var-item"><div class="xr-index-name"><div>__obs__</div></div><div class="xr-index-preview">PandasIndex</div><div></div><input id="index-f708be56-670e-4315-8183-2fa9a4c66a72" class="xr-index-data-in" type="checkbox"><label for="index-f708be56-670e-4315-8183-2fa9a4c66a72" title="Show/Hide index repr"><svg class="icon xr-icon-database"><use href="#icon-database"></use></svg></label><div class="xr-index-data"><pre>PandasIndex(Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
       ...
       909, 910, 911, 912, 913, 914, 915, 916, 917, 918],
      dtype='int64', name='__obs__', length=919))</pre></div></li></ul></div></li><li class="xr-section-item"><input id="section-6eb29ec1-b1c8-4c15-8ada-87f7412d555f" class="xr-section-summary-in" type="checkbox" checked=""><label for="section-6eb29ec1-b1c8-4c15-8ada-87f7412d555f" class="xr-section-summary">Attributes: <span>(6)</span></label><div class="xr-section-inline-details"></div><div class="xr-section-details"><dl class="xr-attrs"><dt><span>created_at :</span></dt><dd>2024-05-26T22:50:19.839116+00:00</dd><dt><span>arviz_version :</span></dt><dd>0.18.0</dd><dt><span>inference_library :</span></dt><dd>pymc</dd><dt><span>inference_library_version :</span></dt><dd>5.15.0+23.g19be124e</dd><dt><span>modeling_interface :</span></dt><dd>bambi</dd><dt><span>modeling_interface_version :</span></dt><dd>0.13.1.dev39+gb7d6a6cb</dd></dl></div></li></ul></div></div><br></div>
                      </ul>
                  </div>
            </li>
            
              </ul>
            </div>
            <style> /* CSS stylesheet for displaying InferenceData objects in jupyterlab.
 *
 */

:root {
  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));
  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));
  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));
  --xr-border-color: var(--jp-border-color2, #e0e0e0);
  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);
  --xr-background-color: var(--jp-layout-color0, white);
  --xr-background-color-row-even: var(--jp-layout-color1, white);
  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);
}

html[theme=dark],
body.vscode-dark {
  --xr-font-color0: rgba(255, 255, 255, 1);
  --xr-font-color2: rgba(255, 255, 255, 0.54);
  --xr-font-color3: rgba(255, 255, 255, 0.38);
  --xr-border-color: #1F1F1F;
  --xr-disabled-color: #515151;
  --xr-background-color: #111111;
  --xr-background-color-row-even: #111111;
  --xr-background-color-row-odd: #313131;
}

.xr-wrap {
  display: block;
  min-width: 300px;
  max-width: 700px;
}

.xr-text-repr-fallback {
  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */
  display: none;
}

.xr-header {
  padding-top: 6px;
  padding-bottom: 6px;
  margin-bottom: 4px;
  border-bottom: solid 1px var(--xr-border-color);
}

.xr-header > div,
.xr-header > ul {
  display: inline;
  margin-top: 0;
  margin-bottom: 0;
}

.xr-obj-type,
.xr-array-name {
  margin-left: 2px;
  margin-right: 10px;
}

.xr-obj-type {
  color: var(--xr-font-color2);
}

.xr-sections {
  padding-left: 0 !important;
  display: grid;
  grid-template-columns: 150px auto auto 1fr 20px 20px;
}

.xr-sections.group-sections {
  grid-template-columns: auto;
}

.xr-section-item {
  display: contents;
}

.xr-section-item input {
  display: none;
}

.xr-section-item input + label {
  color: var(--xr-disabled-color);
}

.xr-section-item input:enabled + label {
  cursor: pointer;
  color: var(--xr-font-color2);
}

.xr-section-item input:enabled + label:hover {
  color: var(--xr-font-color0);
}

.xr-section-summary {
  grid-column: 1;
  color: var(--xr-font-color2);
  font-weight: 500;
}

.xr-section-summary > span {
  display: inline-block;
  padding-left: 0.5em;
}

.xr-section-summary-in:disabled + label {
  color: var(--xr-font-color2);
}

.xr-section-summary-in + label:before {
  display: inline-block;
  content: '►';
  font-size: 11px;
  width: 15px;
  text-align: center;
}

.xr-section-summary-in:disabled + label:before {
  color: var(--xr-disabled-color);
}

.xr-section-summary-in:checked + label:before {
  content: '▼';
}

.xr-section-summary-in:checked + label > span {
  display: none;
}

.xr-section-summary,
.xr-section-inline-details {
  padding-top: 4px;
  padding-bottom: 4px;
}

.xr-section-inline-details {
  grid-column: 2 / -1;
}

.xr-section-details {
  display: none;
  grid-column: 1 / -1;
  margin-bottom: 5px;
}

.xr-section-summary-in:checked ~ .xr-section-details {
  display: contents;
}

.xr-array-wrap {
  grid-column: 1 / -1;
  display: grid;
  grid-template-columns: 20px auto;
}

.xr-array-wrap > label {
  grid-column: 1;
  vertical-align: top;
}

.xr-preview {
  color: var(--xr-font-color3);
}

.xr-array-preview,
.xr-array-data {
  padding: 0 5px !important;
  grid-column: 2;
}

.xr-array-data,
.xr-array-in:checked ~ .xr-array-preview {
  display: none;
}

.xr-array-in:checked ~ .xr-array-data,
.xr-array-preview {
  display: inline-block;
}

.xr-dim-list {
  display: inline-block !important;
  list-style: none;
  padding: 0 !important;
  margin: 0;
}

.xr-dim-list li {
  display: inline-block;
  padding: 0;
  margin: 0;
}

.xr-dim-list:before {
  content: '(';
}

.xr-dim-list:after {
  content: ')';
}

.xr-dim-list li:not(:last-child):after {
  content: ',';
  padding-right: 5px;
}

.xr-has-index {
  font-weight: bold;
}

.xr-var-list,
.xr-var-item {
  display: contents;
}

.xr-var-item > div,
.xr-var-item label,
.xr-var-item > .xr-var-name span {
  background-color: var(--xr-background-color-row-even);
  margin-bottom: 0;
}

.xr-var-item > .xr-var-name:hover span {
  padding-right: 5px;
}

.xr-var-list > li:nth-child(odd) > div,
.xr-var-list > li:nth-child(odd) > label,
.xr-var-list > li:nth-child(odd) > .xr-var-name span {
  background-color: var(--xr-background-color-row-odd);
}

.xr-var-name {
  grid-column: 1;
}

.xr-var-dims {
  grid-column: 2;
}

.xr-var-dtype {
  grid-column: 3;
  text-align: right;
  color: var(--xr-font-color2);
}

.xr-var-preview {
  grid-column: 4;
}

.xr-var-name,
.xr-var-dims,
.xr-var-dtype,
.xr-preview,
.xr-attrs dt {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding-right: 10px;
}

.xr-var-name:hover,
.xr-var-dims:hover,
.xr-var-dtype:hover,
.xr-attrs dt:hover {
  overflow: visible;
  width: auto;
  z-index: 1;
}

.xr-var-attrs,
.xr-var-data {
  display: none;
  background-color: var(--xr-background-color) !important;
  padding-bottom: 5px !important;
}

.xr-var-attrs-in:checked ~ .xr-var-attrs,
.xr-var-data-in:checked ~ .xr-var-data {
  display: block;
}

.xr-var-data > table {
  float: right;
}

.xr-var-name span,
.xr-var-data,
.xr-attrs {
  padding-left: 25px !important;
}

.xr-attrs,
.xr-var-attrs,
.xr-var-data {
  grid-column: 1 / -1;
}

dl.xr-attrs {
  padding: 0;
  margin: 0;
  display: grid;
  grid-template-columns: 125px auto;
}

.xr-attrs dt,
.xr-attrs dd {
  padding: 0;
  margin: 0;
  float: left;
  padding-right: 10px;
  width: auto;
}

.xr-attrs dt {
  font-weight: normal;
  grid-column: 1;
}

.xr-attrs dt:hover span {
  display: inline-block;
  background: var(--xr-background-color);
  padding-right: 10px;
}

.xr-attrs dd {
  grid-column: 2;
  white-space: pre-wrap;
  word-break: break-all;
}

.xr-icon-database,
.xr-icon-file-text2 {
  display: inline-block;
  vertical-align: middle;
  width: 1em;
  height: 1.5em !important;
  stroke-width: 0;
  stroke: currentColor;
  fill: currentColor;
}
.xr-wrap{width:700px!important;} </style>
</div>
</div>
<div id="cell-84" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>partial_pooling_model.predict(partial_pooling_results, kind<span class="op">=</span><span class="st">"pps"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Stack chains and draws. pps stands for posterior predictive samples</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>pps <span class="op">=</span> az.extract_dataset(partial_pooling_results, group<span class="op">=</span><span class="st">"posterior_predictive"</span>)[<span class="st">"log_radon"</span>].values</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>pps_df <span class="op">=</span> pd.DataFrame(data<span class="op">=</span>pps).assign(county<span class="op">=</span>df[<span class="st">"county"</span>])</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>y_pred <span class="op">=</span> pps_df.groupby(<span class="st">"county"</span>).mean().mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>y_sample <span class="op">=</span> df.groupby(<span class="st">"county"</span>)[<span class="st">"log_radon"</span>].mean()</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">7</span>))</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a>sns.regplot(x<span class="op">=</span>y_sample, y<span class="op">=</span>y_pred, ax<span class="op">=</span>ax)</span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a>ax.axline(xy1<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, label<span class="op">=</span><span class="st">"diagonal"</span>)</span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>ax.axhline(y<span class="op">=</span>y_pred.mean(), color<span class="op">=</span><span class="st">"C3"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, label<span class="op">=</span><span class="st">"predicted global mean"</span>)</span>
<span id="cb50-14"><a href="#cb50-14" aria-hidden="true" tabindex="-1"></a>ax.legend(loc<span class="op">=</span><span class="st">"lower right"</span>)</span>
<span id="cb50-15"><a href="#cb50-15" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(</span>
<span id="cb50-16"><a href="#cb50-16" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"log(Radon + 0.1) Mean per County (Partial Pooling Model)"</span>,</span>
<span id="cb50-17"><a href="#cb50-17" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span><span class="st">"observed (sample)"</span>,</span>
<span id="cb50-18"><a href="#cb50-18" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span><span class="st">"prediction"</span>,</span>
<span id="cb50-19"><a href="#cb50-19" aria-hidden="true" tabindex="-1"></a>    xlim<span class="op">=</span>(<span class="fl">0.3</span>, <span class="fl">2.7</span>),</span>
<span id="cb50-20"><a href="#cb50-20" aria-hidden="true" tabindex="-1"></a>    ylim<span class="op">=</span>(<span class="fl">0.3</span>, <span class="fl">2.7</span>),</span>
<span id="cb50-21"><a href="#cb50-21" aria-hidden="true" tabindex="-1"></a>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/tomas/Desktop/OSS/bambinos/bambi/bambi/models.py:845: FutureWarning: 'pps' has been replaced by 'response' and is not going to work in the future
  warnings.warn(
/tmp/ipykernel_45847/3145587883.py:4: FutureWarning: extract_dataset has been deprecated, please use extract
  pps = az.extract_dataset(partial_pooling_results, group="posterior_predictive")["log_radon"].values</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-35-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Note that in this case the points are not concentrated along the diagonal (as it was the case for the unpooled model). The reason is that in the partial pooling model the hyperprior shrinks the predictions towards the global mean, represented by the horizonital dashed line.</p>
<p><strong>What does <em>partial pooling</em> exactly mean here?</strong></p>
<p>We said the first model we built performed a <strong>complete pooling</strong> because estimates pooled observations regardless to which county they belong to. We could see that in the coefficients for the <code>floor</code> variable. The estimate for each level was the sample mean for each level, plus prior smoothing, without making any special distinction to observations from different counties.</p>
<p>Then, when we built our second model we said it performed <strong>no pooling</strong>. This was the opposite scenario. Estimates for effects involving a specific county were not informed at all by the information in the other counties.</p>
<p>Now, we say this model performs <strong>partial pooling</strong>. But what does it mean? Well, if we had <strong>complete</strong> pooling and <strong>no</strong> pooling, this must be some type of compromise in between.</p>
<p>In this model, we have a global intercept <span class="math inline">\(\alpha\)</span>, which represents the mean of the response variable across all counties. We also have group-specific intercepts <span class="math inline">\(u_j\)</span> that represent deviations from the global mean specific to each county <span class="math inline">\(j\)</span>. Thess group-specific intercepts are assigned a Normal prior centered at 0. The standard deviations of these priors are considered random, instead of fixed. Since they are random, they are assigned a prior distribution, which is a hyperprior in this case because it is a prior on top of a prior. And that hyperprior is the same distribution for all the county-specific intercepts. Because of that, these random deviations from the global mean are not independent. Indeed, the shared hyperprior is what causes the <strong>partial pooling</strong> in the model estimates. In other words, some information is shared between counties when computing estimates for their effects and it results in a shrinkage towards the global mean.</p>
<p>Connecting what we’ve just said with the figure above we can see the <strong>partial pooling is a compromise between complete pooling (global mean) and no pooling (diagonal)</strong>.</p>
</section>
</section>
<section id="county-specific-intercepts-and-common-predictors" class="level3">
<h3 class="anchored" data-anchor-id="county-specific-intercepts-and-common-predictors">County-specific intercepts and common predictors</h3>
<p>Next, we add the <code>floor</code> global feature (i.e.&nbsp;does not depend on the <code>county</code>) into the model above. We remove the global intercept so Bambi keeps one coefficient for each <code>floor</code> level.</p>
<p>In the original PyMC example, this model is introduced under the <a href="https://docs.pymc.io/projects/examples/en/latest/case_studies/multilevel_modeling.html#varying-intercept-model">Varying intercept model</a> title. We feel that “County-specific intercepts and common predictors” is a more accurate representation of the model we build in Bambi. It is correct to say this is a varying intercept model, because of the county-specific intercepts, but so was the last model we built.</p>
<section id="model-3" class="level4">
<h4 class="anchored" data-anchor-id="model-3">Model</h4>
<p><span class="math display">\[
y = \beta_j + u_k + \varepsilon
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
y &amp;= \text{Response for the (log) radon measurement } \\
\beta_j &amp;= \text{Coefficient for the floor level } j \\
u_k &amp;= \text{Intercept specific to the county } k \\
\varepsilon &amp; = \text{Residual random error}
\end{aligned}
\]</span></p>
</section>
<section id="priors-3" class="level4">
<h4 class="anchored" data-anchor-id="priors-3">Priors</h4>
<section id="common-effects-3" class="level5">
<h5 class="anchored" data-anchor-id="common-effects-3">Common effects</h5>
<p>The common effect in this model is the floor term <span class="math inline">\(\beta_j\)</span></p>
<p><span class="math display">\[
\beta_j \sim \text{Normal}(0, \sigma_{\beta_j})
\]</span></p>
<p>for all <span class="math inline">\(j: 1, 2\)</span> and <span class="math inline">\(\sigma_{\beta_j}\)</span> is a positive constant that we set to <span class="math inline">\(10\)</span>.</p>
</section>
<section id="group-specific-effects-1" class="level5">
<h5 class="anchored" data-anchor-id="group-specific-effects-1">Group-specific effects</h5>
<p><span class="math display">\[
u_k \sim \text{Normal}(0, \sigma_u)
\]</span></p>
<p>for all <span class="math inline">\(j:1, \cdots, 85\)</span>. The hyperprior is</p>
<p>$$</p>
<p>_u ()</p>
<p>$$</p>
<p>and <span class="math inline">\(\tau\)</span> is a positive constant that we set to <span class="math inline">\(1\)</span>.</p>
</section>
<section id="residual-error-3" class="level5">
<h5 class="anchored" data-anchor-id="residual-error-3">Residual error</h5>
<p><span class="math display">\[
\begin{aligned}
\varepsilon &amp; \sim \text{Normal}(0, \sigma) \\
\sigma &amp; \sim \text{Exponential}(\lambda)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\lambda\)</span> is a positive constant that we set to <span class="math inline">\(1\)</span>.</p>
</section>
<section id="notes-1" class="level5">
<h5 class="anchored" data-anchor-id="notes-1">Notes</h5>
<p><span class="math inline">\(\beta_j\)</span> and <span class="math inline">\(u_k\)</span> may look similar. The difference is that the latter is a hierarchical effect (it has a hyperprior), while the former is not.</p>
<div id="cell-91" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>varying_intercept_priors <span class="op">=</span> {</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"floor"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1|county"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span>bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>)),</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sigma"</span>: bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>varying_intercept_model <span class="op">=</span> bmb.Model(</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>    formula<span class="op">=</span><span class="st">"log_radon ~ 0 + floor + (1|county)"</span>,</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>    priors<span class="op">=</span>varying_intercept_priors,</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>    noncentered<span class="op">=</span><span class="va">False</span></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a> )</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>varying_intercept_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="35">
<pre><code>       Formula: log_radon ~ 0 + floor + (1|county)
        Family: gaussian
          Link: mu = identity
  Observations: 919
        Priors: 
    target = mu
        Common-level effects
            floor ~ Normal(mu: 0.0, sigma: 10.0)
        
        Group-level effects
            1|county ~ Normal(mu: 0.0, sigma: Exponential(lam: 1.0))
        
        Auxiliary parameters
            sigma ~ Exponential(lam: 1.0)</code></pre>
</div>
</div>
<div id="cell-92" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>varying_intercept_results <span class="op">=</span> varying_intercept_model.fit()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [sigma, floor, 1|county_sigma, 1|county]</code></pre>
</div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"60a2d7f859074784aa41a418cfa7a222","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sampling 2 chains for 1_000 tune and 1_000 draw iterations (2_000 + 2_000 draws total) took 7 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details</code></pre>
</div>
</div>
<p>When looking at the graph representation of the model we still see the hierarchical structure for the <code>county</code> varying intercepts, but we do not see it for the <code>floor</code> feature as expected.</p>
<div id="cell-94" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>varying_intercept_model.graph()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="37">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-38-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Let us visualize the posterior distributions:</p>
<div id="cell-96" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>az.plot_trace(data<span class="op">=</span>varying_intercept_results, compact<span class="op">=</span><span class="va">True</span>, chain_prop<span class="op">=</span>{<span class="st">"ls"</span>: <span class="st">"-"</span>})<span class="op">;</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>plt.suptitle(<span class="st">"Varying Intercepts Model Trace"</span>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-39-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="varying-intercept-and-slope-model" class="level3">
<h3 class="anchored" data-anchor-id="varying-intercept-and-slope-model">Varying intercept and slope model</h3>
<p>Next we want to include a hierarchical structure in the floor effect.</p>
<section id="model-4" class="level4">
<h4 class="anchored" data-anchor-id="model-4">Model</h4>
<p><span class="math display">\[
y = \beta_j + b_{jk} + \varepsilon
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
y &amp;= \text{Response for the (log) radon measurement}\\
\beta_j &amp;= \text{Coefficient for the floor level } j \\
b_{jk} &amp;= \text{Coefficient for the floor level } j \text{ specific to the county } k\\
\varepsilon &amp; = \text{Residual random error}
\end{aligned}
\]</span></p>
</section>
<section id="priors-4" class="level4">
<h4 class="anchored" data-anchor-id="priors-4">Priors</h4>
<section id="common-effects-4" class="level5">
<h5 class="anchored" data-anchor-id="common-effects-4">Common effects</h5>
<p>The common effect in this model is the floor term <span class="math inline">\(\beta_j\)</span></p>
<p><span class="math display">\[
\beta_j \sim \text{Normal}(0, \sigma_{\beta_j})
\]</span></p>
<p>where <span class="math inline">\(\sigma_{\beta_j}\)</span> is a positive constant that we set to <span class="math inline">\(10\)</span>.</p>
</section>
<section id="group-specific-effects-2" class="level5">
<h5 class="anchored" data-anchor-id="group-specific-effects-2">Group-specific effects</h5>
<p>Here, again, we have the floor effects</p>
<p><span class="math display">\[
b_{jk} \sim \text{Normal}(0, \sigma_{b_j})
\]</span></p>
<p>for <span class="math inline">\(j:1, 2\)</span> and <span class="math inline">\(k: 1, \cdots, 85\)</span>.</p>
<p>The hyperprior is</p>
<p><span class="math display">\[
\sigma_{b_j} \sim \text{Exponential}(\tau)
\]</span></p>
<p>for <span class="math inline">\(j:1, 2\)</span>.</p>
<p><span class="math inline">\(\tau\)</span> is a positive constant that we set to <span class="math inline">\(1\)</span>.</p>
</section>
<section id="residual-error-4" class="level5">
<h5 class="anchored" data-anchor-id="residual-error-4">Residual error</h5>
<p><span class="math display">\[
\begin{aligned}
\varepsilon &amp; \sim \text{Normal}(0, \sigma) \\
\sigma &amp; \sim \text{Exponential}(\lambda)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\lambda\)</span> is a positive constant that we set to 1.</p>
</section>
<section id="notes-2" class="level5">
<h5 class="anchored" data-anchor-id="notes-2">Notes</h5>
<p>Both <span class="math inline">\(\beta_j\)</span> and <span class="math inline">\(b_{jk}\)</span> are floor effects. The difference is that the first one is a common effect, while the second is a group-specific effect. In other words, the second floor effect varies from county to county. These effects represent the county specific deviations from the common floor effect <span class="math inline">\(\beta_j\)</span>. Because of the hyperprior, the <span class="math inline">\(b_{jk}\)</span> effects aren’t independent and result in the partial-pooling effect.</p>
<p>In this case the Bambi model specification is quite easy, namely <code>log_radon ~ 0 + floor + (0 + floor|county)</code>. This formula represents the following terms:</p>
<ul>
<li>The first <code>0</code> tells we don’t want a global intercept.</li>
<li><code>floor</code> is <span class="math inline">\(\beta_j\)</span>. It says we want to include an effect for each floor level. Since there’s no global intercept, a coefficient for each level is included.</li>
<li>The <code>0</code> in <code>(0 + floor|county)</code> means we don’t want county-specific intercept. We need to explicitly turn it off as we did with the regular intercept.</li>
<li><code>floor|county</code> is <span class="math inline">\(b_{jk}\)</span>, the county-specific floor coefficients. Again, since there’s no varying intercepot for the counties, this includes coefficients for both floor levels.</li>
</ul>
<div id="cell-103" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>varying_intercept_slope_priors <span class="op">=</span> {</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"floor"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"floor|county"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span>bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>)),</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sigma"</span>: bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>varying_intercept_slope_model <span class="op">=</span> bmb.Model(</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>    formula<span class="op">=</span><span class="st">"log_radon ~ 0 + floor + (0 + floor|county)"</span>,</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>    priors<span class="op">=</span>varying_intercept_slope_priors,</span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>    noncentered<span class="op">=</span><span class="va">True</span></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a> )</span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>varying_intercept_slope_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="39">
<pre><code>       Formula: log_radon ~ 0 + floor + (0 + floor|county)
        Family: gaussian
          Link: mu = identity
  Observations: 919
        Priors: 
    target = mu
        Common-level effects
            floor ~ Normal(mu: 0.0, sigma: 10.0)
        
        Group-level effects
            floor|county ~ Normal(mu: 0.0, sigma: Exponential(lam: 1.0))
        
        Auxiliary parameters
            sigma ~ Exponential(lam: 1.0)</code></pre>
</div>
</div>
<p>Next, we fit the model. Note we increase the default number of <code>draws</code> from the posterior and the <code>tune</code> samples to 2000. In addition, as the structure of the model gets more complex, so does the posterior. That’s why we increase <code>target_accept</code> from the default 0.8 to 0.9, because we want to explore the posterior more cautiously .</p>
<div id="cell-105" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>varying_intercept_slope_results <span class="op">=</span> varying_intercept_slope_model.fit(</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>    draws<span class="op">=</span><span class="dv">2000</span>, </span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>    tune<span class="op">=</span><span class="dv">2000</span>,</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>    target_accept<span class="op">=</span><span class="fl">0.9</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [sigma, floor, floor|county_sigma, floor|county_offset]</code></pre>
</div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"1f791094fc294104ab894757656b5165","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sampling 2 chains for 2_000 tune and 2_000 draw iterations (4_000 + 4_000 draws total) took 27 seconds.
We recommend running at least 4 chains for robust computation of convergence diagnostics</code></pre>
</div>
</div>
<p>In the graph representation of the model we can now see hierarchical structures both in the intercepts and the slopes. The terms that end with <code>_offset</code> appeared because we are using a non-centered parametrization. This parametrization is an algebraic trick that helps computation but leaves the model unchanged.</p>
<div id="cell-107" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>varying_intercept_slope_model.graph()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-42-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Let’s have a look at the marginal posterior for the coefficients in the model.</p>
<div id="cell-109" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>var_names <span class="op">=</span> [<span class="st">"floor"</span>, <span class="st">"floor|county"</span>, <span class="st">"floor|county_sigma"</span>, <span class="st">"sigma"</span>]</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>az.plot_trace(</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>varying_intercept_slope_results,</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>    var_names<span class="op">=</span>var_names, </span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>    compact<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>    chain_prop<span class="op">=</span>{<span class="st">"ls"</span>: <span class="st">"-"</span>}</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-43-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="adding-group-level-predictors" class="level3">
<h3 class="anchored" data-anchor-id="adding-group-level-predictors">Adding group-level predictors</h3>
<p>We now want to consider a county-level predictor, namely the (log) uranium level. This is not a county-level predictor in the sense that we use a county-specific coefficient, but in the sense that all the uranium concentrations were measured per county. Thus all the houses in the same county have the same uranium level.</p>
<section id="model-5" class="level4">
<h4 class="anchored" data-anchor-id="model-5">Model</h4>
<p><span class="math display">\[
y = \beta_j + \xi x + b_{jk} + \varepsilon
\]</span></p>
<p>where</p>
<p><span class="math display">\[
\begin{aligned}
y &amp;= \text{Response for the (log) radon measurement} \\
x &amp;= \text{Log uranium concentration} \\
\beta_j &amp;= \text{Coefficient for the floor level } j \\
\xi &amp;= \text{Coefficient for the slope of the log uranium concentration}\\
b_{jk} &amp;= \text{Coefficient for the floor level } j \text{ specific to the county } k\\
\varepsilon &amp; = \text{Residual random error}
\end{aligned}
\]</span></p>
</section>
<section id="priors-5" class="level4">
<h4 class="anchored" data-anchor-id="priors-5">Priors</h4>
<section id="common-effects-5" class="level5">
<h5 class="anchored" data-anchor-id="common-effects-5">Common effects</h5>
<p>This model has two common effects:</p>
<p><span class="math display">\[
\begin{aligned}
\beta_j \sim \text{Normal}(0, \sigma_{\beta_j}) \\
\xi \sim \text{Normal}(0, \sigma_\xi)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(j:1, 2\)</span> and all <span class="math inline">\(\sigma_{\beta_j}\)</span> and <span class="math inline">\(\sigma_{\xi}\)</span> are set to <span class="math inline">\(10\)</span>.</p>
</section>
<section id="group-specific-effects-3" class="level5">
<h5 class="anchored" data-anchor-id="group-specific-effects-3">Group-specific effects</h5>
<p>Here, again, we have the floor effects</p>
<p><span class="math display">\[
b_{jk} \sim \text{Normal}(0, \sigma_{b_j})
\]</span></p>
<p>for <span class="math inline">\(j:1, 2\)</span> and <span class="math inline">\(k: 1, \cdots, 85\)</span>.</p>
<p>The hyperprior is</p>
<p><span class="math display">\[
\sigma_{b_j} \sim \text{Exponential}(\tau)
\]</span></p>
<p>for <span class="math inline">\(j:1, 2\)</span>.</p>
<p><span class="math inline">\(\tau\)</span> is a positive constant that we set to <span class="math inline">\(1\)</span>.</p>
</section>
<section id="residual-error-5" class="level5">
<h5 class="anchored" data-anchor-id="residual-error-5">Residual error</h5>
<p><span class="math display">\[
\begin{aligned}
\varepsilon &amp; \sim \text{Normal}(0, \sigma) \\
\sigma &amp; \sim \text{Exponential}(\lambda)
\end{aligned}
\]</span></p>
<p>where <span class="math inline">\(\lambda\)</span> is a positive constant that we set to <span class="math inline">\(1\)</span>.</p>
<div id="cell-114" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>covariate_priors <span class="op">=</span> {</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"floor"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"log_u"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span><span class="dv">10</span>),</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"floor|county"</span>: bmb.Prior(<span class="st">"Normal"</span>, mu<span class="op">=</span><span class="dv">0</span>, sigma<span class="op">=</span>bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>)),</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"sigma"</span>: bmb.Prior(<span class="st">"Exponential"</span>, lam<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>covariate_model <span class="op">=</span> bmb.Model(</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>    formula<span class="op">=</span><span class="st">"log_radon ~ 0 + floor + log_u + (0 + floor|county)"</span>,</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>    priors<span class="op">=</span>covariate_priors,</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>    noncentered<span class="op">=</span><span class="va">True</span></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a> )</span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>covariate_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="43">
<pre><code>       Formula: log_radon ~ 0 + floor + log_u + (0 + floor|county)
        Family: gaussian
          Link: mu = identity
  Observations: 919
        Priors: 
    target = mu
        Common-level effects
            floor ~ Normal(mu: 0.0, sigma: 10.0)
            log_u ~ Normal(mu: 0.0, sigma: 10.0)
        
        Group-level effects
            floor|county ~ Normal(mu: 0.0, sigma: Exponential(lam: 1.0))
        
        Auxiliary parameters
            sigma ~ Exponential(lam: 1.0)</code></pre>
</div>
</div>
<div id="cell-115" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>covariate_results <span class="op">=</span> covariate_model.fit(</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    draws<span class="op">=</span><span class="dv">2000</span>, </span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    tune<span class="op">=</span><span class="dv">2000</span>,</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    target_accept<span class="op">=</span><span class="fl">0.9</span>,</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    chains<span class="op">=</span><span class="dv">2</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (2 chains in 2 jobs)
NUTS: [sigma, floor, log_u, floor|county_sigma, floor|county_offset]</code></pre>
</div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"65eee87d7ed34e228581177723723e8b","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Sampling 2 chains for 2_000 tune and 2_000 draw iterations (4_000 + 4_000 draws total) took 39 seconds.
There were 3 divergences after tuning. Increase `target_accept` or reparameterize.
We recommend running at least 4 chains for robust computation of convergence diagnostics</code></pre>
</div>
</div>
<div id="cell-116" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>covariate_model.graph()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="45">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-46-output-1.svg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-117" class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>var_names <span class="op">=</span> [<span class="st">"floor"</span>, <span class="st">"log_u"</span>, <span class="st">"floor|county"</span>, <span class="st">"floor|county_sigma"</span>, <span class="st">"sigma"</span>]</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>az.plot_trace(</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>covariate_results,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>    var_names<span class="op">=</span>var_names, </span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>    compact<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>    chain_prop<span class="op">=</span>{<span class="st">"ls"</span>: <span class="st">"-"</span>}</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-47-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Let us now visualize the posterior distributions of the intercepts:</p>
<div id="cell-119" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get log_u values per county</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>log_u_sample <span class="op">=</span> df.groupby([<span class="st">"county"</span>])[<span class="st">"log_u"</span>].mean().values</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="co"># compute the slope posterior samples</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>log_u_slope <span class="op">=</span> covariate_results.posterior[<span class="st">"log_u"</span>].values[..., <span class="va">None</span>] <span class="op">*</span> log_u_sample</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the posterior for the floor coefficient when it is Basement</span></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>intercepts <span class="op">=</span> (</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>    covariate_results.posterior.sel(floor_dim<span class="op">=</span><span class="st">"Basement"</span>)[<span class="st">"floor"</span>]</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> covariate_results.posterior.sel(floor__expr_dim<span class="op">=</span><span class="st">"Basement"</span>)[<span class="st">"floor|county"</span>] </span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>).values</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>y_predicted <span class="op">=</span> (intercepts <span class="op">+</span> log_u_slope).reshape(<span class="dv">4000</span>, n_counties).T</span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a><span class="co"># reduce the intercepts posterior samples to the mean per county</span></span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a>mean_intercept <span class="op">=</span> intercepts.mean(axis<span class="op">=</span><span class="dv">2</span>)[..., <span class="va">None</span>] <span class="op">+</span> log_u_slope</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-120" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>y_predicted_bounds <span class="op">=</span> np.quantile(y_predicted, q<span class="op">=</span>[<span class="fl">0.03</span>, <span class="fl">0.96</span>], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>log_u_sample,</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>y_predicted.mean(axis<span class="op">=</span><span class="dv">1</span>),</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>    alpha<span class="op">=</span><span class="fl">0.8</span>,</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">"C0"</span>,</span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>    s<span class="op">=</span><span class="dv">50</span>,</span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a>    label<span class="op">=</span><span class="st">"Mean county-intercept"</span>,</span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a>    ax<span class="op">=</span>ax</span>
<span id="cb74-13"><a href="#cb74-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb74-14"><a href="#cb74-14" aria-hidden="true" tabindex="-1"></a>ax.vlines(log_u_sample, y_predicted_bounds[<span class="dv">0</span>], y_predicted_bounds[<span class="dv">1</span>], color<span class="op">=</span><span class="st">"C1"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb74-15"><a href="#cb74-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-16"><a href="#cb74-16" aria-hidden="true" tabindex="-1"></a>az.plot_hdi(</span>
<span id="cb74-17"><a href="#cb74-17" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>log_u_sample,</span>
<span id="cb74-18"><a href="#cb74-18" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>mean_intercept,</span>
<span id="cb74-19"><a href="#cb74-19" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">"black"</span>,</span>
<span id="cb74-20"><a href="#cb74-20" aria-hidden="true" tabindex="-1"></a>    fill_kwargs<span class="op">=</span>{<span class="st">"alpha"</span>: <span class="fl">0.1</span>, <span class="st">"label"</span>: <span class="st">"Mean intercept HPD"</span>},</span>
<span id="cb74-21"><a href="#cb74-21" aria-hidden="true" tabindex="-1"></a>    ax<span class="op">=</span>ax</span>
<span id="cb74-22"><a href="#cb74-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb74-23"><a href="#cb74-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-24"><a href="#cb74-24" aria-hidden="true" tabindex="-1"></a>sns.lineplot(</span>
<span id="cb74-25"><a href="#cb74-25" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span>log_u_sample,</span>
<span id="cb74-26"><a href="#cb74-26" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>mean_intercept.reshape(<span class="dv">4000</span>, n_counties).mean(axis<span class="op">=</span><span class="dv">0</span>),</span>
<span id="cb74-27"><a href="#cb74-27" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">"black"</span>,</span>
<span id="cb74-28"><a href="#cb74-28" aria-hidden="true" tabindex="-1"></a>    alpha<span class="op">=</span><span class="fl">0.6</span>,</span>
<span id="cb74-29"><a href="#cb74-29" aria-hidden="true" tabindex="-1"></a>    label<span class="op">=</span><span class="st">"Mean intercept"</span>,</span>
<span id="cb74-30"><a href="#cb74-30" aria-hidden="true" tabindex="-1"></a>    ax<span class="op">=</span>ax</span>
<span id="cb74-31"><a href="#cb74-31" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb74-32"><a href="#cb74-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-33"><a href="#cb74-33" aria-hidden="true" tabindex="-1"></a>ax.legend(loc<span class="op">=</span><span class="st">"upper left"</span>)</span>
<span id="cb74-34"><a href="#cb74-34" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(</span>
<span id="cb74-35"><a href="#cb74-35" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"County Intercepts (Covariance Model)"</span>,</span>
<span id="cb74-36"><a href="#cb74-36" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span><span class="st">"County-level log uranium"</span>,</span>
<span id="cb74-37"><a href="#cb74-37" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span><span class="st">"Intercept estimate"</span></span>
<span id="cb74-38"><a href="#cb74-38" aria-hidden="true" tabindex="-1"></a>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-49-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
</section>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>Let us dig deeper into the model comparison for the pooled, unpooled, and partial pooling models. To do so we are generate predictions for each model ad county level, where we aggregate by taking the mean, and plot them against the observed values.</p>
<div id="cell-123" class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate posterior predictive samples</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>pooled_model.predict(pooled_results, kind<span class="op">=</span><span class="st">"response"</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>unpooled_model.predict(unpooled_results, kind<span class="op">=</span><span class="st">"response"</span>)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>partial_pooling_model.predict(partial_pooling_results, kind<span class="op">=</span><span class="st">"response"</span>)</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a><span class="co"># stack chain and draw values</span></span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a>pooled_pps <span class="op">=</span> az.extract_dataset(pooled_results, group<span class="op">=</span><span class="st">"posterior_predictive"</span>)[<span class="st">"log_radon"</span>].values</span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a>unpooled_pps <span class="op">=</span> az.extract_dataset(unpooled_results, group<span class="op">=</span><span class="st">"posterior_predictive"</span>)[<span class="st">"log_radon"</span>].values</span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a>partial_pooling_pps <span class="op">=</span> az.extract_dataset(partial_pooling_results, group<span class="op">=</span><span class="st">"posterior_predictive"</span>)[<span class="st">"log_radon"</span>].values</span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate predictions per county</span></span>
<span id="cb75-12"><a href="#cb75-12" aria-hidden="true" tabindex="-1"></a>pooled_pps_df <span class="op">=</span> pd.DataFrame(data<span class="op">=</span>pooled_pps).assign(county<span class="op">=</span>df[<span class="st">"county"</span>])</span>
<span id="cb75-13"><a href="#cb75-13" aria-hidden="true" tabindex="-1"></a>y_pred_pooled <span class="op">=</span> pooled_pps_df.groupby(<span class="st">"county"</span>).mean().mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb75-14"><a href="#cb75-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-15"><a href="#cb75-15" aria-hidden="true" tabindex="-1"></a>unpooled_pps_df <span class="op">=</span> pd.DataFrame(data<span class="op">=</span>unpooled_pps).assign(county<span class="op">=</span>df[<span class="st">"county"</span>])</span>
<span id="cb75-16"><a href="#cb75-16" aria-hidden="true" tabindex="-1"></a>y_pred_unpooled <span class="op">=</span> unpooled_pps_df.groupby(<span class="st">"county"</span>).mean().mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb75-17"><a href="#cb75-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-18"><a href="#cb75-18" aria-hidden="true" tabindex="-1"></a>partial_pooling_pps_df <span class="op">=</span> pd.DataFrame(data<span class="op">=</span>partial_pooling_pps).assign(county<span class="op">=</span>df[<span class="st">"county"</span>])</span>
<span id="cb75-19"><a href="#cb75-19" aria-hidden="true" tabindex="-1"></a>y_pred_partial_pooling <span class="op">=</span> partial_pooling_pps_df.groupby(<span class="st">"county"</span>).mean().mean(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb75-20"><a href="#cb75-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-21"><a href="#cb75-21" aria-hidden="true" tabindex="-1"></a><span class="co"># observed values</span></span>
<span id="cb75-22"><a href="#cb75-22" aria-hidden="true" tabindex="-1"></a>y_sample <span class="op">=</span> df.groupby(<span class="st">"county"</span>)[<span class="st">"log_radon"</span>].mean()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_45847/1610299561.py:7: FutureWarning: extract_dataset has been deprecated, please use extract
  pooled_pps = az.extract_dataset(pooled_results, group="posterior_predictive")["log_radon"].values
/tmp/ipykernel_45847/1610299561.py:8: FutureWarning: extract_dataset has been deprecated, please use extract
  unpooled_pps = az.extract_dataset(unpooled_results, group="posterior_predictive")["log_radon"].values
/tmp/ipykernel_45847/1610299561.py:9: FutureWarning: extract_dataset has been deprecated, please use extract
  partial_pooling_pps = az.extract_dataset(partial_pooling_results, group="posterior_predictive")["log_radon"].values</code></pre>
</div>
</div>
<div id="cell-124" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">8</span>))</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>sns.regplot(x<span class="op">=</span>y_sample, y<span class="op">=</span>y_pred_pooled, label<span class="op">=</span><span class="st">"pooled"</span>, color<span class="op">=</span><span class="st">"C0"</span>, ax<span class="op">=</span>ax)</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>sns.regplot(x<span class="op">=</span>y_sample, y<span class="op">=</span>y_pred_unpooled, label<span class="op">=</span><span class="st">"unpooled"</span>, color<span class="op">=</span><span class="st">"C1"</span>, ax<span class="op">=</span>ax)</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>sns.regplot(x<span class="op">=</span>y_sample, y<span class="op">=</span>y_pred_partial_pooling, label<span class="op">=</span><span class="st">"partial pooling"</span>, color<span class="op">=</span><span class="st">"C2"</span>, ax<span class="op">=</span>ax)</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>ax.axhline(y<span class="op">=</span>df[<span class="st">"log_radon"</span>].mean(), color<span class="op">=</span><span class="st">"C0"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, label<span class="op">=</span><span class="st">"sample mean"</span>)</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>ax.axline(xy1<span class="op">=</span>(<span class="dv">1</span>, <span class="dv">1</span>), slope<span class="op">=</span><span class="dv">1</span>, color<span class="op">=</span><span class="st">"black"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, label<span class="op">=</span><span class="st">"diagonal"</span>)</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>ax.axhline(</span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span>y_pred_partial_pooling.mean(), color<span class="op">=</span><span class="st">"C3"</span>,</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>    linestyle<span class="op">=</span><span class="st">"--"</span>, label<span class="op">=</span><span class="st">"predicted global mean (partial pooling)"</span></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>ax.legend(loc<span class="op">=</span><span class="st">"upper center"</span>, bbox_to_anchor<span class="op">=</span>(<span class="fl">0.5</span>, <span class="op">-</span><span class="fl">0.1</span>), ncol<span class="op">=</span><span class="dv">2</span>)</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>ax.<span class="bu">set</span>(</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span><span class="st">"log(Radon + 0.1) Mean per County - Model Comparison"</span>,</span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>    xlabel<span class="op">=</span><span class="st">"observed (sample)"</span>,</span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a>    ylabel<span class="op">=</span><span class="st">"prediction"</span>,</span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a>    xlim<span class="op">=</span>(<span class="fl">0.2</span>, <span class="fl">2.8</span>),</span>
<span id="cb77-18"><a href="#cb77-18" aria-hidden="true" tabindex="-1"></a>    ylim<span class="op">=</span>(<span class="fl">0.2</span>, <span class="fl">2.8</span>),</span>
<span id="cb77-19"><a href="#cb77-19" aria-hidden="true" tabindex="-1"></a>)<span class="op">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="radon_example_files/figure-html/cell-51-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>The pooled model consider all the counties together, this explains why the predictions do not vary at county level. This is represented by the almost-flat line in the plot above (blue).</li>
<li>On the other hand, the unpooled model considers each county separately, so the prediction is very close to the observation mean. This is represented by the line very close to the diagonal (orange).</li>
<li>The partial pooling model is mixing global and information at county level. This is clearly seen by how corresponding (green) line is in between the pooling and unpooling lines.</li>
</ul>
<div id="cell-126" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>load_ext watermark</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>watermark <span class="op">-</span>n <span class="op">-</span>u <span class="op">-</span>v <span class="op">-</span>iv <span class="op">-</span>w</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Last updated: Sun May 26 2024

Python implementation: CPython
Python version       : 3.11.9
IPython version      : 8.24.0

seaborn   : 0.13.2
matplotlib: 3.8.4
pymc      : 5.15.0+23.g19be124e
bambi     : 0.13.1.dev39+gb7d6a6cb
numpy     : 1.26.4
pandas    : 2.2.2
arviz     : 0.18.0

Watermark: 2.4.3
</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../notebooks/sleepstudy.html" class="pagination-link" aria-label="Hierarchical Linear Regression (Sleepstudy example)">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Hierarchical Linear Regression (Sleepstudy example)</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../notebooks/Strack_RRR_re_analysis.html" class="pagination-link" aria-label="Bayesian Workflow (Strack RRR Analysis Replication)">
        <span class="nav-page-text">Bayesian Workflow (Strack RRR Analysis Replication)</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© Copyright 2024, The developers of Bambi.</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>